---
title: "homer_rankplot_iterate.rmd"
author: "Jrose"
date: "1/26/2021"
output: html_document
---

Here I am looking to refine the original heatmap of homer enrichment ranks scores generated by homer_rankplots.R

Ideas to acomplish:

- Exclude factors which are not expressed in RNA data
- Add side annotation of un-Zscored average rank value
- Change to diverging color scheme

## Task 1: Exclude factors not expressed in RNA

```{r}
library(tidyverse)
library(here)

input_dir <- "output"
output_dir <- "output"

load(here(input_dir, "Tmem_StimUnstim_TFrank.rda"))

str(data_nodups)
```

Loading RNA data

```{r}
library(readr)

RNA_input_dir <- 'input_RNA'

RNA <- read_csv(here(RNA_input_dir, "Tcell_RNA.combined.RPKM.csv"), col_names=TRUE)
#str(RNA)
files = read.table(here(RNA_input_dir, "RNAseq.sample.manifest.txt"), sep = "\t", header = T, as.is = T)
files = files[grepl("EM|CM|Nav", files$group),]
#files = files[!grepl("CD4_EMRA", files$group),]
files = files[files$include,]
#str(files)
```

Removing genes with expression lower than set cutoff value by group using bistools filter_deteted function

```{r}
#Filter_detected function from bistools
filter_detected = function(x, group, cutoff){

	one_gene = data.frame(group=group, rpm=as.vector(t(x)))
	one_gene = aggregate(rpm~group, data=one_gene, FUN=function(x) sum(x>cutoff))

	if( any( one_gene$rpm >= table(group) ) )
		{ return(TRUE) }
	else
		{ return(FALSE) }
}

cutoff <- 3
expr_test = apply(RNA[,grep("(CD8|CD4)_((CM)|(EM)|(EMRA)|(N))(Stim|).rpkm", colnames(RNA))], 1, function(x) filter_detected(x, files$group, cutoff) )

RNA_sub <- RNA[expr_test,]

rpkm<- cbind(RNA_sub[,c("SYMBOL")], RNA_sub[,grep("(CD8|CD4)_((CM)|(EM)|(EMRA)|(N))(Stim|).rpkm", colnames(RNA_sub))])
colnames(rpkm )[1] <- "SYMBOL"
#str(rpkm)

```

df of mean rpkm values for use later

```{r}
mean_rpkm  <- pivot_longer(rpkm , 
                          cols=ends_with('.rpkm'),
                          names_to=c("Sample", "Celltype", "Subtype"),
                          names_pattern= "RNA_(.*)_(.*)_(.*).rpkm",
                          values_to="rpkm"
                          ) %>%
  mutate(cellSubtype=paste(Celltype, Subtype, sep="_")) %>%
  group_by(SYMBOL, cellSubtype) %>%
  summarise(mean_rpkm=mean(rpkm)) %>%
  pivot_wider(names_from=cellSubtype, values_from=mean_rpkm) %>%
  select(SYMBOL, CD4_N, CD4_NStim, CD4_CM, CD4_CMStim, CD4_EM, CD4_EMStim, CD8_N, CD8_NStim, CD8_CM, CD8_CMStim, CD8_EM, CD8_EMStim, CD8_EMRA, CD8_EMRAStim)

str(mean_rpkm)
```

Can I find matches between TF name and gene name??

Note to self: I think I need to use a grep statement here instead of %in%

```{r}
tomatch <- unique(RNA_sub$SYMBOL)
#tomatch <- gsub("-|\\.", "", tomatch)
supp <- c("NFAT", "HIF-1a", "NRF", "CEBP")
tomatch <- c(tomatch, supp)

# grep(paste(tomatch,collapse="|"), data_nodups$TF2, ignore.case = T)
# 
# grep(paste(unique(data_nodups$TF2),collapse="|"), RNA$SYMBOL, ignore.case = T)

#It seems that regex patterns can't be longer than ~2554 gene long! so I have to do this monstrocity!
index <- c(grep(paste(tomatch[1:2500],collapse="|"), data_nodups$TF2, ignore.case = T),grep(paste(tomatch[2501:5054],collapse="|"), data_nodups$TF2, ignore.case = T),grep(paste(tomatch[5055:7607],collapse="|"), data_nodups$TF2, ignore.case = T),grep(paste(tomatch[7608:10160],collapse="|"), data_nodups$TF2, ignore.case = T),grep(paste(tomatch[10160:10415],collapse="|"), data_nodups$TF2, ignore.case = T))
index <- unique(index)

data_TF_RNAcheck <- data_nodups[index,]
length(unique(data_TF_RNAcheck$TF))
unique(data_TF_RNAcheck$TF)[order(unique(data_TF_RNAcheck$TF))]
```

It looks like I can get ~180 of the 389 motifs confirmed to their RNA data

Junk code I used to figure this out:
```{r}
table(data_nodups$TF2 %in% RNA$SYMBOL)
table(data_nodups$TF2 %in% RNA_sub$SYMBOL)
data_nodups$TF[data_nodups$TF2 %in% RNA_sub$SYMBOL]
head(RNA_sub$SYMBOL)
head(data_nodups$TF2)
length(unique(data_nodups$TF))
unique(data_nodups$TF2)

table(unique(data_nodups$TF2) %in% RNA$SYMBOL)

RNA_sub$SYMBOL[grep(paste(unique(data_nodups$TF2),collapse="|"), RNA_sub$SYMBOL, ignore.case=T)]

grep("AP1", RNA$SYMBOL, ignore.case=T)

RNA$SYMBOL[grep("AP1", RNA$SYMBOL, ignore.case=T)]

head(data_nodups)
names(RNA_sub)

data_nodups$TF[grep("NFAT", data_nodups$TF)]
mean_rpkm[grep("NFAT", mean_rpkm$SYMBOL),]
tomatch[grep("NFAT", tomatch)]
tomatch[grep("HIF", tomatch)]
data_nodups$TF[grep("HIF", data_nodups$TF)]
data_nodups$TF[grep("NRF", data_nodups$TF)]
data_nodups$TF[grep("AP1", data_nodups$TF)]
data_nodups$TF[grep("S1P", data_nodups$TF2, ignore.case = T)]
data_nodups$TF[grep("MYC", data_nodups$TF2, ignore.case = T)]
data_nodups$TF[grep("sP", data_nodups$TF2, ignore.case = T)]
data_nodups$TF[grep("LA", data_nodups$TF2, ignore.case = T)]
```

Let's try one more time with a fuzzy string matching package: stringdist

```{r}
library(stringdist)

amatch("HIF1A", data_nodups$TF2, maxDist=2)
data_nodups$TF2[569]

amatch("^NFATC1", data_nodups$TF2, maxDist=2)
data_nodups$TF2[294]

amatch("^DUSP1", data_nodups$TF2, maxDist=2)
data_nodups$TF2[258]

agrep("^HIF1A", data_nodups$TF2, max.distance = 3)
data_nodups$TF2[agrep("^HIF1A", data_nodups$TF2, max.distance = 2.5)]

data_nodups$TF2[agrep("^DUSP1", data_nodups$TF2, max.distance = 2.5)]
```

Not quite good enough.


# Task 2: Add side annotation of un-Zscored average rank value

```{r}
library(ComplexHeatmap)

col = colorRamp2(c(-2, 0, 2), c("forest green", "white", "purple"), space = "sRGB")
col2 = colorRamp2(c(-2, 0, 2), c("tan4", "white", "deepskyblue3"), space = "sRGB")
name="Enrichment Rank (z-scored)"

highlight = c("BATF(bZIP)", "IRF4(IRF)", "Ets1-distal(ETS)", "bZIP:IRF(bZIP,IRF)", "Ap4(bHLH)", "Tcf21(bHLH)", "MyoD(bHLH)", "MyoG(bHLH)",
              "Bach1(bZIP)", "RUNX2(Runt)", "RUNX1(Runt)", "Tbet(T-box)", "Nrf2(bZIP)", "IRF8(IRF)", "RORg(NR)", "IRF3(IRF)",
              "NFAT:AP1(RHD,bZIP)", "NFkB-p50,p52(RHD)","CTCF(Zf)", "JunB(bZIP)", "LEF1(HMG)", "Tcf7(HMG)",
              "GATA(Zf),IR4", "RUNX(Runt)", "PRDM1(Zf)", "Stat3(Stat)", "STAT5(Stat)", "Bach2(bZIP)")
highlight2 = gsub("\\(", "\\\\(", highlight)
#^List of factor motifs to highlight

lbl_index <- vector()
for (i in 1:length(highlight)){
  if (highlight[i]%in%rownames(mtx_rank3)) {
    lbl_index[i] <- grep(highlight2[i], rownames(mtx_rank3))
  }
}

Heatmap(mtx_rank3,
        #show_row_names = F,
        row_names_gp = gpar(fontsize = 4),
        col=col,
        name=name
        #right_annotation = rowAnnotation(gene = anno_mark(at = lbl_index, 
        #                                                  labels = highlight,
        #                                                  labels_gp = gpar(fontsize=10)))
        )


Heatmap(mtx_rank3[,-4],
        show_row_names = F,
        #row_names_gp = gpar(fontsize = 4),
        right_annotation = rowAnnotation(gene = anno_mark(at = lbl_index,
                                                          labels = highlight,
                                                          labels_gp = gpar(fontsize=10))),
        col=col2,
        name=name
)

cutoff <- apply(mtx_rank, c(1), function(x) any(x>5))
mtx_rank2 <- mtx_rank[cutoff,]

mtx_rank2_sub <- mtx_rank2[lbl_index,]
mtx_rank2_sub <- select(mtx_rank3, -stim)
mtx_rank3_sub <- t(apply(mtx_rank2_sub, c(1), scale))
mtx_rank3_sub <- drop_na(data.frame(mtx_rank3_sub))
colnames(mtx_rank3_sub) <- colnames(mtx_rank2_sub)

Heatmap(mtx_rank3_sub,
        show_row_names = T,
        #row_names_gp = gpar(fontsize = 4),
        # right_annotation = rowAnnotation(gene = anno_mark(at = lbl_index,
        #                                                   labels = highlight,
        #                                                   labels_gp = gpar(fontsize=10))),
        col=col,
        name=name
)
```

# Task 3:

Let's try and create heatmaps of % in peakset vs background for just a few examples of each TF group

```{r}
newdat <- data_nodups %>% group_by(TF, cluster) %>% summarise(ave_per_target = mean(Percent.of.Target.Sequences.with.Motif), ave_per_background =mean(Percent.of.Background.Sequences.with.Motif), ave_per_diff = mean(per_diff), P.value=mean(P.value), ave_Npeaks=mean(Number.of.Target.Sequences.with.Motif), n=n()) #%>% 
  #subset(TF=="LEF1(HMG)" | TF=="IRF8(IRF)" | TF=="NFAT:AP1(RHD,bZIP)")

newdat <- newdat[newdat$TF%in%c("LEF1(HMG)", "IRF8(IRF)", "NFAT:AP1(RHD,bZIP)"),]

newdat$cluster <- factor(newdat$cluster)
newdat$cluster <- recode(newdat$cluster, both="Augmented", poised="Induced", rest = "Expressed")
newdat <- subset(newdat, cluster!="stim")
newdat
```

```{r}
ggplot(newdat, aes(x=cluster, y=ave_per_target, fill=TF)) + geom_bar(position="dodge", stat="identity")
```

heatmap

```{r}
tar_back <- select(newdat, TF, cluster, ave_per_target, ave_per_background) %>% pivot_wider(names_from="cluster", values_from=c("ave_per_target", "ave_per_background")) %>% select(TF, ends_with("Expressed"), ends_with("Augmented"), ends_with("Induced"), -starts_with("ave_per_background"))

mtx <- as.matrix(tar_back[,-1])
rownames(mtx) <- tar_back$TF
mtx_scale <- t(apply(mtx, c(1), scale))
rownames(mtx_scale) <- tar_back$TF
colnames(mtx_scale) <- colnames(mtx)
mtx_scale
mtx
```

```{r}
library(ComplexHeatmap)
library(circlize)

#col3 = colorRamp2(c(0, 15, 30), c("navy", "purple", "firebrick"), space = "sRGB")

Heatmap(mtx_scale, cluster_rows = T, cluster_columns = T)
Heatmap(mtx, cluster_rows = T, cluster_columns = F, #col=col3
        )
```

```{r}
newdat2 <- data_nodups %>% group_by(TF_family, cluster) %>% summarise(ave_per_target = mean(Percent.of.Target.Sequences.with.Motif), ave_per_background =mean(Percent.of.Background.Sequences.with.Motif), ave_per_diff = mean(per_diff), P.value=mean(P.value), ave_Npeaks=mean(Number.of.Target.Sequences.with.Motif), n=n()) #%>% 
  #subset(TF=="LEF1(HMG)" | TF=="IRF8(IRF)" | TF=="NFAT:AP1(RHD,bZIP)")

newdat2 <- newdat2[newdat2$TF_family%in%c("HMG", "IRF", "RHD", "bZIP"),]

newdat2$cluster <- factor(newdat2$cluster)
newdat2$cluster <- recode(newdat2$cluster, both="Augmented", poised="Induced", rest = "Expressed")
newdat2 <- subset(newdat2, cluster!="stim")
newdat2
```
```{r}
ggplot(newdat2, aes(x=cluster, y=ave_per_target , fill=TF_family)) + geom_bar(position="dodge", stat="identity")
```

