---
title: "ATAC_stim_category_RNA.Rmd"
author: "Jrose"
date: "12/18/2020"
output: html_document
---

In this log I'm going to try and intersect the ATAC peaks which have been sorted by categories in the ATAC_peak_cat.R script with the RNA data of different stimulation expression programs (induced, expressed, etc)

The output of the ATAC_peak_cat.R has also been filtered for detectable accessibility regions using the ATAC_peak_cat_trim.R script before being analyzed here.

```{r message=FALSE, warning=FALSE}
setwd("~/Documents/Emory/Boss/data/ATAC")

library(tidyverse)
library(readr)
library(here)

input_dir <- "Stim_Unstim/diff"
output_dir <- "Stim_Unstim/atac_peak_cat/output"

diff <- read_tsv(here(input_dir, "diff.glm.CD4_EM_cat_trim.ATACseq.v1_1.txt"), col_names=TRUE)

diff <- mutate(diff, peak=factor(peak), simpleAnnot=factor(simpleAnnot), CD4_CM_peak_cat=factor(CD4_CM_peak_cat),CD4_EM_peak_cat=factor(CD4_EM_peak_cat),CD4_EMRA_peak_cat=factor(CD4_EMRA_peak_cat), CD8_CM_peak_cat=factor(CD8_CM_peak_cat), CD8_EM_peak_cat=factor(CD8_EM_peak_cat), CD8_EMRA_peak_cat=factor(CD8_EMRA_peak_cat))


#Gettign dataframes of stim state
load("/home/jim/Documents/Emory/Boss/data/RNA/StimUnstim/stim_unstim_all_full_folds.rda")
```


Extracting out the RNA categories

```{r}
#Poised genes
poised = map(full_folds, function(x) subset(x, sig=="Stim_sig"))

for (i in 1:length(names(poised))){
  poised[[i]]$subset = names(poised)[i]
}

poised_df <- Reduce(
  function(x, y, ...) merge(x, y, all = TRUE, ...),
  poised
)
poised_df$sig_full <- factor(poised_df$sig_full)
poised_max_df <- subset(poised_df, Stim_logFC>1.5)

#Both genes
both = map(full_folds, function(x) subset(x, sig=="Both"))

for (i in 1:length(names(both))){
  both[[i]]$subset = names(both)[i]
}

both_df <- Reduce(
  function(x, y, ...) merge(x, y, all = TRUE, ...),
  both
)
both_df$sig_full <- factor(both_df$sig_full)
both_max_df <- subset(both_df, Stim_logFC>1.5)

#Resting genes
rest = map(full_folds, function(x) subset(x, sig=="Unstim_sig"))

for (i in 1:length(names(rest))){
  rest[[i]]$subset = names(rest)[i]
}

rest_df <- Reduce(
  function(x, y, ...) merge(x, y, all = TRUE, ...),
  rest
)
rest_df$sig_full <- factor(rest_df$sig_full)
rest_max_df <- subset(rest_df, Stim_logFC>1.5)

#Need to generate unique identifiers for subset inclusion
rest_subs <- select(rest_df, ENTREZID, subset, sig_full) %>% group_by(ENTREZID) %>%
  summarize(Expr_subsets = paste(subset, collapse=":"))
poi_subs <- select(poised_df, ENTREZID, subset, sig_full) %>% group_by(ENTREZID) %>%
  summarize(Ind_subsets = paste(subset, collapse=":"))
both_subs <- select(both_df, ENTREZID, subset, sig_full) %>% group_by(ENTREZID) %>%
  summarize(Aug_subsets = paste(subset, collapse=":"))

subs <- full_join(rest_subs, poi_subs, by="ENTREZID")
subs <- full_join(subs, both_subs, by="ENTREZID")
head(subs)
```

Combining peaks and gene categories

```{r}
ELSE <- TRUE

ATAC <- left_join(diff, subs, by=c("Entrez.ID"="ENTREZID")) %>% 
  #subset(!is.na(Expr_subsets) | !is.na(Ind_subsets) | !is.na(Aug_subsets)) %>% 
  mutate(CD8_Induced = case_when(
                          grepl("CD8", Ind_subsets) ~ TRUE,
                          ELSE ~ FALSE
                                ),
         CD4_Induced = case_when(
                          grepl("CD4", Ind_subsets) ~ TRUE,
                          ELSE ~ FALSE
                                ),
         CD8_Expressed = case_when(
                          grepl("CD8", Expr_subsets) ~ TRUE,
                          ELSE ~ FALSE
                                ),
         CD4_Expressed = case_when(
                          grepl("CD4", Expr_subsets) ~ TRUE,
                          ELSE ~ FALSE
                                ),
         CD8_Augmented = case_when(
                          grepl("CD8", Aug_subsets) ~ TRUE,
                          ELSE ~ FALSE
                                ),
         CD4_Augmented = case_when(
                          grepl("CD4", Aug_subsets) ~ TRUE,
                          ELSE ~ FALSE
                                ),
         CD8_Category = case_when(
                        CD8_Augmented==T & CD8_Induced==F & CD8_Expressed==F ~ "Augmented",
                        CD8_Augmented==F & CD8_Induced==T & CD8_Expressed==F ~ "Induced",
                        CD8_Augmented==F & CD8_Induced==F & CD8_Expressed==T ~ "Expressed",
                        CD8_Augmented==T & CD8_Induced==T & CD8_Expressed==F ~ "Aug/Ind",
                        CD8_Augmented==T & CD8_Induced==F & CD8_Expressed==T ~ "Aug/Exp",
                        CD8_Augmented==F & CD8_Induced==T & CD8_Expressed==T ~ "Exp/Ind",
                        CD8_Augmented==F & CD8_Induced==F & CD8_Expressed==F ~ "Unexpressed",
                        ELSE ~"Unknown"
                                ),
         CD4_Category = case_when(
                        CD4_Augmented==T & CD4_Induced==F & CD4_Expressed==F ~ "Augmented",
                        CD4_Augmented==F & CD4_Induced==T & CD4_Expressed==F ~ "Induced",
                        CD4_Augmented==F & CD4_Induced==F & CD4_Expressed==T ~ "Expressed",
                        CD4_Augmented==T & CD4_Induced==T & CD4_Expressed==F ~ "Aug/Ind",
                        CD4_Augmented==T & CD4_Induced==F & CD4_Expressed==T ~ "Aug/Exp",
                        CD4_Augmented==F & CD4_Induced==T & CD4_Expressed==T ~ "Exp/Ind",
                        CD4_Augmented==F & CD4_Induced==F & CD4_Expressed==F ~ "Unexpressed",
                        ELSE ~ "Unknown"
                                ),
  )


ATAC$CD8_Category <- factor(ATAC$CD8_Category)
ATAC$CD4_Category <- factor(ATAC$CD4_Category)

summary(ATAC$CD4_Category)
summary(ATAC$CD8_Category)
summary(ATAC$CD4_CM_peak_cat)
summary(diff$CD4_CM_peak_cat)
```

Lots of NAs was surprising. Looked into it and those are CD4/CD8 specific 

```{r}
library(ggpubr)

overall_peak_n <- summary(ATAC$CD4_EM_peak_cat)

# CD4_induced <- subset(ATAC, CD4_Category=="Induced" | CD4_Category=="Aug/Ind" | CD4_Category == "Exp/Ind")
# CD4_induced_n <- summary(CD4_induced$CD4_EM_peak_cat)
# CD4_Augmented <- subset(ATAC, CD4_Category=="Augmented" | CD4_Category=="Aug/Ind" | CD4_Category == "Au/Exp")
# CD4_Augmented_n <- summary(CD4_Augmented$CD4_EM_peak_cat)
# CD4_Expressed <- subset(ATAC, CD4_Category=="Expressed" | CD4_Category=="Aug/Exp" | CD4_Category == "Exp/Ind")
# CD4_Expressed_n <- summary(CD4_Expressed$CD4_EM_peak_cat)

CD4_induced <- subset(ATAC, CD4_Category=="Induced")
CD4_induced_n <- summary(CD4_induced$CD4_EM_peak_cat)
CD4_Augmented <- subset(ATAC, CD4_Category=="Augmented")
CD4_Augmented_n <- summary(CD4_Augmented$CD4_EM_peak_cat)
CD4_Expressed <- subset(ATAC, CD4_Category=="Expressed")
CD4_Expressed_n <- summary(CD4_Expressed$CD4_EM_peak_cat)

per_fnc <- function(x) {
  per <- round((x/sum(x))*100, 2)
}


pie_df <- data.frame(peaktype = names(overall_peak_n), 
                     Overall_n = overall_peak_n, 
                     Overall_per = per_fnc(overall_peak_n),
                     CD4_Induced_n = CD4_induced_n,
                     CD4_Induced_per = per_fnc(CD4_induced_n),
                     CD4_Augmented_n = CD4_Augmented_n,
                     CD4_Augmented_per = per_fnc(CD4_Augmented_n),
                     CD4_Expressed_n = CD4_Expressed_n,
                     CD4_Expressed_per = per_fnc(CD4_Expressed_n)
                     )

reform_df <- select(pie_df, peaktype, ends_with("per")) %>% pivot_longer(cols=ends_with("per"), names_to="Gene_Expression", values_to="Percent")
reform_df$Gene_Expression <- gsub("_per", "", reform_df$Gene_Expression)

reform_df_n <- select(pie_df, peaktype, ends_with("_n")) %>% pivot_longer(cols=ends_with("_n"), names_to="Gene_Expression", values_to="Count")
reform_df_n$Gene_Expression <- gsub("_n", "", reform_df_n$Gene_Expression)

ggbarplot(reform_df, x = "peaktype", y = "Percent",
               fill = "Gene_Expression", 
               palette = c("CD4_Augmented" = "purple",
                           "CD4_Expressed"="firebrick",
                           "CD4_Induced"="orange",
                           "Overall"="black"),
               x.text.angle= 90,
               add = c("mean_sd"), 
               position = position_dodge(width=0.8)
              )
ggsave(filename = here(output_dir, "CD4_peaktype_by_expression_bar.pdf"), height=7.5, width=8)

ggbarplot(reform_df_n, x = "peaktype", y = "Count",
               fill = "Gene_Expression", 
               palette = "jco",
               x.text.angle= 90,
               add = c("mean_sd"), 
               position = position_dodge(width=0.8)
              )

ggbarplot(reform_df, x = "peaktype", y = "Percent",
               fill = "Gene_Expression", 
               palette = c("CD4_Augmented" = "purple",
                           "CD4_Expressed"="firebrick",
                           "CD4_Induced"="orange",
                           "Overall"="black"),
               x.text.angle= 90,
               add = c("mean_sd"), 
               position = position_dodge(width=0.8)
              ) + coord_flip() + 
              ylab("Percent of Peaks in Cis")
ggsave(filename = here(output_dir, "CD4_peaktype_by_expression_bar_flip.pdf"), height=4, width=8)
```

What about where these types of peaks are located in each with the trimmed.

```{r}
annot_all <- diff %>% group_by(simpleAnnot) %>% summarise(n=n()) %>% mutate(CD4_EM_peak_cat = "All Peaks", sum=sum(n), per=n/sum)

annot_plot <- ATAC %>% group_by(CD4_EM_peak_cat, simpleAnnot) %>% summarise(n=n())
peak_cat_sum <- annot_plot %>% ungroup() %>% group_by(CD4_EM_peak_cat) %>% summarise(sum=sum(n))
peak_cat_sum
annot_plot <- left_join(annot_plot, peak_cat_sum)
annot_plot <- mutate(annot_plot, per=n/sum)

annot_plot2 <- bind_rows(annot_plot, annot_all)
ggplot(annot_plot2, aes(x="", y=per, fill=simpleAnnot)) + geom_bar(stat="identity", width=1) +
  coord_polar("y", start=0) + 
  facet_wrap(vars(CD4_EM_peak_cat)) +
  geom_text(aes(label = ifelse(simpleAnnot=="Promoter" | simpleAnnot=="Intergenic" | simpleAnnot=="Intron", paste0(round((per*100), 2), "%"), "")), position = position_stack(vjust=0.5), size=3, min.segment.length = 0) +
  theme_void()
ggsave(filename = here(output_dir, "CD4_EM_peaktype_annotation_pies.pdf"), height=7, width=8)
```

Would be easier to first make some bed files for this I think:

Need a color scheme for each category:

Conserved: 51,24,50    #331832
Induced: 192, 87, 70   #C05746
Primed: 175, 160, 96  #AFA060
Memory: 103,87,158    #67579E
                  Old color: Memory: 40, 83, 107   #28536B
Naive: 42,157,143     #2A9D8F
                  Old color: Naive: 251, 177, 60
Other: 131,139,139   #838B8B
                  Old color Other: 79, 93, 47   #4F5D2F

```{r}
custom_pal <- c("#331832","#C05746", "#AFA060", "#67579E", "#2A9D8F", "#838B8B")
names(custom_pal) <- c("Conserved", "Induced", "Primed", "Memory", "Naive", "Other")
pie(rep(1,length(custom_pal)),labels=names(custom_pal),col=custom_pal, radius=1)
```



```{r}
bed_dir <- "Stim_Unstim/diff/bed"

ATAC %>% select(chr, start, end, peak, strand, ends_with("cat")) %>% mutate(score=0, thickStart=start, thickEnd=end)

vars <- names(ATAC)[grep("peak_cat", names(ATAC))]

for (i in 1:length(vars)){
  data <- ATAC %>% select(chr, start, end, peak, strand, ends_with("cat")) %>%   
    mutate(score=0, thickStart=start, thickEnd=end,
           peak_cat = paste(peak, .data[[vars[i]]], sep="_"), 
           itemRgb = case_when( .data[[vars[i]]] =="Conserved" ~ "51,24,50",
                               .data[[vars[i]]] =="Induced" ~ "192,87,70",
                               .data[[vars[i]]] =="Primed" ~ "175,160,96",
                               .data[[vars[i]]] =="Memory" ~ "103,87,158",
                               .data[[vars[i]]] =="Naive" ~ "42,157,143",
                               .data[[vars[i]]] =="Other" ~ "131,139,139"
                              )                                                                     
          ) %>% select(chr, start, end, peak_cat, score, strand, thickStart, thickEnd, itemRgb)
  write.table(data, file=here(bed_dir, paste(vars[i], ".bed", sep="")), quote=F, sep="\t", row.names=F, col.names=F)
   write.table(data[grep("Conserved", data$peak_cat),], file=here(bed_dir, paste(vars[i], "_consv.bed", sep="")), quote=F, sep="\t", row.names=F, col.names=F)
   write.table(data[grep("Induced", data$peak_cat),], file=here(bed_dir, paste(vars[i], "_induc.bed", sep="")), quote=F, sep="\t", row.names=F, col.names=F)
  write.table(data[grep("Primed", data$peak_cat),], file=here(bed_dir, paste(vars[i], "_prime.bed", sep="")), quote=F, sep="\t", row.names=F, col.names=F)
  write.table(data[grep("Memory", data$peak_cat),], file=here(bed_dir, paste(vars[i], "_mem.bed", sep="")), quote=F, sep="\t", row.names=F, col.names=F)
  write.table(data[grep("Naive", data$peak_cat),], file=here(bed_dir, paste(vars[i], "_nav.bed", sep="")), quote=F, sep="\t", row.names=F, col.names=F)
  write.table(data[grep("Other", data$peak_cat),], file=here(bed_dir, paste(vars[i], "_other.bed", sep="")), quote=F, sep="\t", row.names=F, col.names=F)
}
```

### Naive category is rare. Where are these at?

```{r}
subset(ATAC, CD4_EM_peak_cat =="Naive") %>% arrange(desc(CD8_Nav_blood_unstim.v.CD8_EM_blood_unstim.logFC)) %>% select(peak, Gene.Name, CD4_EM_peak_cat, Detailed.Annotation, ends_with("cat"), ends_with("subsets"))
```

Let's look fro examples of each of the peaks at key genes

```{r}
Gene = "CCR7"

subset(ATAC, Gene.Name==Gene) %>% arrange(start) %>% select(peak, CD4_EM_peak_cat, Gene.Name, Detailed.Annotation, ends_with("cat"), ends_with("subsets")) 

subset(ATAC, Gene.Name==Gene) %>% select(peak, Gene.Name, Detailed.Annotation, ends_with("cat"), ends_with("subsets")) %>% group_by(CD4_EM_peak_cat) %>% summarize(cat_n = n())
```


### Memory

```{r}
subset(ATAC, CD4_EM_peak_cat =="Memory") %>% arrange(CD4_Nav_blood_unstim.v.CD4_EM_blood_unstim.logFC) %>% select(peak, Gene.Name, CD4_EM_peak_cat, Detailed.Annotation, CD4_Nav_blood_unstim.v.CD4_EM_blood_unstim.logFC,ends_with("cat"), ends_with("subsets"))
```



### Now let's look at primed

```{r}
subset(ATAC, CD4_EM_peak_cat =="Primed") %>% arrange(CD4_Nav_blood_unstim.v.CD4_EM_blood_unstim.logFC) %>% select(peak, Gene.Name, CD4_EM_peak_cat, Detailed.Annotation, ends_with("cat"), ends_with("subsets"))
```

```{r}
Gene = "IL7R"

subset(ATAC, Gene.Name==Gene) %>% arrange(start) %>% select(peak, CD4_EM_peak_cat, Gene.Name, Detailed.Annotation, ends_with("cat"), ends_with("subsets")) 

subset(ATAC, Gene.Name==Gene) %>% select(peak, Gene.Name, Detailed.Annotation, ends_with("cat"), ends_with("subsets")) %>% group_by(CD4_EM_peak_cat) %>% summarize(cat_n = n())
```

What about primed peaks that are also at promoters?

Hypothesis: These will be enriched at inducible genes

```{r}
prime_promoters <- subset(ATAC, CD4_EM_peak_cat=="Primed" & simpleAnnot =="Promoter") %>% arrange(CD4_Nav_blood_unstim.v.CD4_EM_blood_unstim.logFC) %>% select(peak, Gene.Name, CD4_EM_peak_cat, CD4_Category,CD8_Category)
prime_promoters
summary(prime_promoters$CD4_Category)
summary(prime_promoters$CD8_Category)
```

Intersting genes in this group. It seems to be more enriched for expressed genes though. Perhaps I need to rethink how things are labeled?

What If I turn it around and do inducible at promoters?

```{r}
Ind_promoters <- subset(ATAC, CD4_EM_peak_cat=="Induced" & simpleAnnot =="Promoter") %>% arrange(CD4_Nav_blood_unstim.v.CD4_Nav_blood_stim.logFC) %>% select(peak, Gene.Name, CD4_EM_peak_cat, CD4_Category,CD8_Category)
Ind_promoters
summary(Ind_promoters$CD4_Category)
summary(Ind_promoters$CD8_Category)
```

A bit more induced category genes here including IL5, IL31 but also a fair amount of expressed genes too. 
  Does this group include things that are downregulated after stimulation??
  
May also have to do some stats on this to really know for sure.

Are there any genes with the other types of peaks at their promoters?
```{r}
Mem_promoters <- subset(ATAC, CD4_EM_peak_cat=="Memory" & simpleAnnot =="Promoter") %>% arrange(CD4_Nav_blood_unstim.v.CD4_EM_blood_unstim.logFC) %>% select(peak, Gene.Name, CD4_EM_peak_cat, CD4_Category,CD8_Category)
Mem_promoters
```

```{r}
Nav_promoters <- subset(ATAC, CD4_EM_peak_cat=="Naive" & simpleAnnot =="Promoter") %>% arrange(CD4_Nav_blood_unstim.v.CD4_EM_blood_unstim.logFC) %>% select(peak, Gene.Name, CD4_EM_peak_cat, CD4_Category,CD8_Category)
Nav_promoters
```


### Discrete heatmap of peak categories

There are too many peaks to visualize in a single heaetmap so I'll start with just promoters

```{r}
promoters <- subset(ATAC, simpleAnnot=="Promoter") %>% select(peak, ends_with("cat"), -CD4_EMRA_peak_cat)
names(promoters)

prom_mtx <- as.matrix(promoters[,-1])
rownames(prom_mtx) <- promoters$peak
```

```{r}
library(ComplexHeatmap)

custom_pal <- c("#331832","#C05746", "#AFA060", "#67579E", "#2A9D8F", "#838B8B")
names(custom_pal) <- c("Conserved", "Induced", "Primed", "Memory", "Naive", "Other")

Heatmap(prom_mtx, show_row_names = F, col=custom_pal)
```
 Too many conserved promoters! Let's get rid of them
 
```{r}
promoters_noCon <- subset(promoters, CD4_CM_peak_cat !="Conserved" & CD4_EM_peak_cat !="Conserved" & CD8_CM_peak_cat !="Conserved" & CD8_EM_peak_cat !="Conserved") %>% select(peak, CD4_CM_peak_cat, CD4_EM_peak_cat, CD8_CM_peak_cat, CD8_EM_peak_cat, CD8_EMRA_peak_cat)

summary(promoters_noCon)

prom_mtx2 <- as.matrix(promoters_noCon[,-1])
rownames(prom_mtx2) <- promoters_noCon$peak
colnames(prom_mtx2) <- gsub("_peak_cat", "", colnames(prom_mtx2))
```
 
```{r}
library(ComplexHeatmap)

custom_pal <- c("#331832","#C05746", "#AFA060", "#67579E", "#2A9D8F", "#838B8B")
names(custom_pal) <- c("Conserved", "Induced", "Primed", "Memory", "Naive", "Other")

pdf(file=here(output_dir, "promoters_peakcat_heatmap.pdf"), useDingbats = F)
Heatmap(prom_mtx2, show_row_names = F, col=custom_pal)
dev.off()

```

Interesting that there seems to be conservation between CD4 cells and CD8 for primed and memory promoters. ALso interesting to see interply between prmied and induced as well as Naive and other categories between samples.

EM/EMRA differentiated cells have higher number of naive promoter peaks.


### Let's plot out the non-conserved peaks at promoters for various gene expression programs

The concept here is that perhaps mem-induced genes have primed promoters? 

```{r}
ATAC_promoter <- subset(ATAC, simpleAnnot=="Promoter" & CD4_CM_peak_cat !="Conserved" & CD4_EM_peak_cat !="Conserved" & CD8_CM_peak_cat !="Conserved" & CD8_EM_peak_cat !="Conserved")
#ATAC_promoter <- subset(ATAC, simpleAnnot=="Promoter")
ATAC_promoter_GInd <- ATAC_promoter[ATAC_promoter$Entrez.ID %in% poised_df$ENTREZID,] %>% select(peak, Gene.Name, CD4_CM_peak_cat, CD4_EM_peak_cat, CD8_CM_peak_cat, CD8_EM_peak_cat, CD8_EMRA_peak_cat)

ATAC_promoter_GAug <- ATAC_promoter[ATAC_promoter$Entrez.ID %in% both_df$ENTREZID,] %>% select(peak, Gene.Name, CD4_CM_peak_cat, CD4_EM_peak_cat, CD8_CM_peak_cat, CD8_EM_peak_cat, CD8_EMRA_peak_cat)

prom_mtx_GInd <- as.matrix(ATAC_promoter_GInd[,3:7])
rownames(prom_mtx_GInd) <- ATAC_promoter_GInd$Gene.Name
colnames(prom_mtx_GInd) <- gsub("_peak_cat", "", colnames(prom_mtx_GInd))

```

```{r}
library(ComplexHeatmap)

pdf(file=here(output_dir, "Mem_Indc_Promoters_peak_cat_heat.pdf"), useDingbats = F)
Heatmap(prom_mtx_GInd, 
        col=custom_pal, 
        show_row_names = T, 
        row_names_gp = gpar(fontsize=4),
        rect_gp = gpar(col = "white", lwd = 0.5)
        )
dev.off()
```

```{r}
ATAC_promoter_GAug <- ATAC_promoter[ATAC_promoter$Entrez.ID %in% both_df$ENTREZID,] %>% select(peak, Gene.Name, CD4_CM_peak_cat, CD4_EM_peak_cat, CD8_CM_peak_cat, CD8_EM_peak_cat, CD8_EMRA_peak_cat)

prom_mtx_GAug <- as.matrix(ATAC_promoter_GAug[,3:7])
rownames(prom_mtx_GAug) <- ATAC_promoter_GAug$Gene.Name
colnames(prom_mtx_GAug) <- gsub("_peak_cat", "", colnames(prom_mtx_GAug))
```

```{r}
library(ComplexHeatmap)

pdf(file=here(output_dir, "Mem_Aug_Promoters_peak_cat_heat.pdf"), useDingbats = F)
Heatmap(prom_mtx_GAug, 
        col=custom_pal, 
        show_row_names = T, 
        row_names_gp = gpar(fontsize=6),
        rect_gp = gpar(col = "white", lwd = 0.5)
        )
dev.off()
```

```{r}
ATAC_promoter_GExp <- ATAC_promoter[ATAC_promoter$Entrez.ID %in% rest_df$ENTREZID,] %>% select(peak, Gene.Name, CD4_CM_peak_cat, CD4_EM_peak_cat, CD8_CM_peak_cat, CD8_EM_peak_cat, CD8_EMRA_peak_cat)

prom_mtx_GExp <- as.matrix(ATAC_promoter_GExp[,3:7])
rownames(prom_mtx_GExp) <- ATAC_promoter_GExp$Gene.Name
colnames(prom_mtx_GExp) <- gsub("_peak_cat", "", colnames(prom_mtx_GExp))
```

```{r}
library(ComplexHeatmap)

pdf(file=here(output_dir, "Mem_Exp_Promoters_peak_cat_heat.pdf"), useDingbats = F)
Heatmap(prom_mtx_GExp, 
        col=custom_pal, 
        show_row_names = T, 
        row_names_gp = gpar(fontsize=6),
        rect_gp = gpar(col = "white", lwd = 0.5)
        )
dev.off()
```

Hard to say what the difference between the three expression programs are from just these plots. One thing that seems to be clear though is that several 'important' genes happen to have induced or primed peaks at their promoters. 

Maybe there is a better way to visualize this? Perhaps some sort of rank ordering? Need to think on how to do it. 

## Count gene locus in cis with at least one primed peak

The concept here is that even one primed peak at a locus indicates potenial for enhanced expression. I need to:

1. Create a list of all gene locus with at least one primed peak nearby
  - Should distance to TSS be a cutoff?
      Depends on how many I get
2. Look at expression changes after stimulation of these genes
  - Hypothesis: Genes with primed peaks in cis should show enhanced expression in stimulated memory cells.
  -> logFC (MemStim/NaiveStim)
  
### Working on part 1 from above:

```{r}
primed <- subset(ATAC, CD4_CM_peak_cat == "Primed" |
            CD4_EM_peak_cat == "Primed" |
            CD8_CM_peak_cat == "Primed" |
            CD8_EM_peak_cat == "Primed" |
            CD8_EMRA_peak_cat == "Primed"
       )

summary(factor(primed$CD4_CM_peak_cat))
summary(primed$Distance.to.TSS)
ggplot(primed, aes(x=Distance.to.TSS)) + geom_density() + theme_minimal()
ggplot(primed, aes(x=Distance.to.TSS)) + geom_histogram(binwidth=1000) + theme_minimal() + scale_y_continuous(limits=c(0,250)) + scale_x_continuous(limits=c(-5e+05, 5e+05))
ggsave(filename = here(output_dir, "primed_distance_to_TSS_hist.pdf"), width=4, height=3)
```

Okay now I have ~10K peaks that have some primed characterization in some memory subset. It also looks like their distance to TSS is roughly normally distributed which makes sense.

The IQR of dist.to.TSS is between -12kb and +31kb. I could potentially exclude anythign beyond these.

Let's see how many fall into this window:
```{r}
primed_strict <- subset(primed, Distance.to.TSS >= -11992 & Distance.to.TSS < 31083)
```

That knocks it down by about half to 5200 remaining peaks.

How many unique genes are in each of these datasets?
```{r}
length(unique(primed$Gene.Name))
length(unique(primed_strict$Gene.Name))
```

Not terrible. 


#### Let's do the same for memory and other peak types

```{r}
Mem <- subset(ATAC, CD4_CM_peak_cat == "Memory" |
            CD4_EM_peak_cat == "Memory" |
            CD8_CM_peak_cat == "Memory" |
            CD8_EM_peak_cat == "Memory" |
            CD8_EMRA_peak_cat == "Memory"
       )

summary(factor(Mem$CD4_CM_peak_cat))
summary(Mem$Distance.to.TSS)
ggplot(Mem, aes(x=Distance.to.TSS)) + geom_density() + theme_minimal()
ggplot(Mem, aes(x=Distance.to.TSS)) + geom_histogram(binwidth=1000) + theme_minimal() + scale_y_continuous(limits=c(0,250)) + scale_x_continuous(limits=c(-5e+05, 5e+05))
```

```{r}
Ind <- subset(ATAC, CD4_CM_peak_cat == "Induced" |
            CD4_EM_peak_cat == "Induced" |
            CD8_CM_peak_cat == "Induced" |
            CD8_EM_peak_cat == "Induced" |
            CD8_EMRA_peak_cat == "Induced"
       )

summary(factor(Ind$CD4_CM_peak_cat))
summary(Ind$Distance.to.TSS)
ggplot(Ind, aes(x=Distance.to.TSS)) + geom_density() + theme_minimal()
ggplot(Ind, aes(x=Distance.to.TSS)) + geom_histogram(binwidth=1000) + theme_minimal() + scale_y_continuous(limits=c(0,250)) + scale_x_continuous(limits=c(-5e+05, 5e+05))
```

```{r}
Consv <- subset(ATAC, CD4_CM_peak_cat == "Conserved" |
            CD4_EM_peak_cat == "Conserved" |
            CD8_CM_peak_cat == "Conserved" |
            CD8_EM_peak_cat == "Conserved" |
            CD8_EMRA_peak_cat == "Conserved"
       )

summary(factor(Consv$CD4_CM_peak_cat))
summary(Consv$Distance.to.TSS)
ggplot(Consv, aes(x=Distance.to.TSS)) + geom_density() + theme_minimal()
ggplot(Consv, aes(x=Distance.to.TSS)) + geom_histogram(binwidth=1000) + theme_minimal() + scale_y_continuous(limits=c(0,250)) + scale_x_continuous(limits=c(-5e+05, 5e+05))
```

```{r}
Nav <- subset(ATAC, CD4_CM_peak_cat == "Naive" |
            CD4_EM_peak_cat == "Naive" |
            CD8_CM_peak_cat == "Naive" |
            CD8_EM_peak_cat == "Naive" |
            CD8_EMRA_peak_cat == "Naive"
       )

summary(factor(Nav$CD4_CM_peak_cat))
summary(Nav$Distance.to.TSS)
ggplot(Nav, aes(x=Distance.to.TSS)) + geom_density() + theme_minimal()
ggplot(Nav, aes(x=Distance.to.TSS)) + geom_histogram(binwidth=1000) + theme_minimal() + scale_y_continuous(limits=c(0,250)) + scale_x_continuous(limits=c(-5e+05, 5e+05))
```

```{r}
Other <- subset(ATAC, CD4_CM_peak_cat == "Other" |
            CD4_EM_peak_cat == "Other" |
            CD8_CM_peak_cat == "Other" |
            CD8_EM_peak_cat == "Other" |
            CD8_EMRA_peak_cat == "Other"
       )

summary(factor(Other$CD4_CM_peak_cat))
summary(Other$Distance.to.TSS)
ggplot(Other, aes(x=Distance.to.TSS)) + geom_density() + theme_minimal()
ggplot(Other, aes(x=Distance.to.TSS)) + geom_histogram(binwidth=1000) + theme_minimal() + scale_y_continuous(limits=c(0,250)) + scale_x_continuous(limits=c(-5e+05, 5e+05))
```


#### How much to Mem and other peaks overlap?

```{r}
table(primed$Gene.Name %in% Mem$Gene.Name)
table(primed$Gene.Name %in% Ind$Gene.Name)
table(primed$Gene.Name %in% Consv$Gene.Name)
table(primed$Gene.Name %in% Nav$Gene.Name)
```

A lot of overlap here.

### Now let's see if the epxression of the primed peaks is correlated

Now I need to add in the actual expression data to the ATAC dataframe

```{r}
fqDir = "/home/jim/Documents/Emory/Boss/data/"
fqFile = "RNAseq.sample.manifest.txt";
files = read.table(paste0(fqDir, fqFile), sep = "\t", header = T, as.is = T);
files = files[grepl("EM|CM|Nav", files$group),]
files = files[!grepl("CD4_EMRA", files$group),]
files = files[which(files$include),]
#files = files[files$stim=="unstim",]
files = mutate(files, cellSubtype = paste(cellType, cellSubset, stim, sep="_"))

RNA <- read_tsv(file="/home/jim/Documents/Emory/Boss/data/RNA/StimUnstim/diff.significant.glm.Human_Tcell_RNA.txt",col_names=T)

#Subset just significant in desired group comparisons
grps <- unique(files$group)
smpl_comp <- apply(expand.grid(grps, grps), 1, paste, collapse=".v.")
smpl_comp_sig <- paste(smpl_comp, ".sig", sep="")
smpl_comp <- c(paste(smpl_comp, ".logFC", sep=""),
               paste(smpl_comp, ".logCPM", sep=""),
               paste(smpl_comp, ".PValue", sep=""),
               paste(smpl_comp, ".fdr", sep=""),
               paste(smpl_comp, ".sig", sep="")
)
keep <-RNA[apply(RNA[,names(RNA)%in%smpl_comp_sig],1,function(x) any(x==TRUE)),'SYMBOL']
RNA_sub = RNA[RNA$SYMBOL%in%keep$SYMBOL,]

table(RNA_sub$ENTREZID %in% primed$Entrez.ID)
table(RNA_sub$SYMBOL %in% primed$Gene.Name)

ELSE <- TRUE
RNA_sub <- mutate(RNA_sub, PrimeState=case_when(
                              SYMBOL %in% primed$Gene.Name ~ TRUE,
                              ELSE ~ FALSE
                                        ),
                           PrimeState_Strict = case_when(
                             SYMBOL %in% primed_strict$Gene.Name ~ TRUE,
                             ELSE ~ FALSE
                                        ),
                           MemState=case_when(
                              SYMBOL %in% Mem$Gene.Name ~ TRUE,
                              ELSE ~ FALSE
                                        ),
                           IndState=case_when(
                              SYMBOL %in% Ind$Gene.Name ~ TRUE,
                              ELSE ~ FALSE
                                        ),
                           ConsvState=case_when(
                              SYMBOL %in% Consv$Gene.Name ~ TRUE,
                              ELSE ~ FALSE
                                        ),
                           NavState=case_when(
                              SYMBOL %in% Nav$Gene.Name ~ TRUE,
                              ELSE ~ FALSE
                                        ),
                          OtherState=case_when(
                              SYMBOL %in% Other$Gene.Name ~ TRUE,
                              ELSE ~ FALSE
                                        )
                          
            )

summary(RNA_sub$PrimeState)
summary(RNA_sub$PrimeState_Strict)
summary(RNA_sub$MemState)
summary(RNA_sub$IndState)
summary(RNA_sub$ConsvState)
summary(RNA_sub$NavState)
summary(RNA_sub$OtherState)
```

Let's plot logFCs

```{r}
library(ggExtra)

FCcutoff <- 0
#FCcutoff <- 1  #^Switched to no logFC cutoff on 28Feb22 to check
FDRcutoff <- 0.05

RNA_sub2 <- subset(RNA_sub, abs(CD4_Nav_blood_unstim.v.CD4_EM_blood_unstim.logFC)>FCcutoff & CD4_Nav_blood_unstim.v.CD4_EM_blood_unstim.fdr <FDRcutoff | abs(CD4_Nav_blood_stim.v.CD4_EM_blood_stim.logFC)>FCcutoff & CD4_Nav_blood_stim.v.CD4_EM_blood_stim.fdr <FDRcutoff) %>%
  mutate(CD4_EM_blood_unstim.v.CD4_Nav_blood_unstim.logFC=-CD4_Nav_blood_unstim.v.CD4_EM_blood_unstim.logFC,
                   CD4_EM_blood_stim.v.CD4_Nav_blood_stim.logFC=-CD4_Nav_blood_stim.v.CD4_EM_blood_stim.logFC
                   )

p <- ggplot(RNA_sub2, aes(x=CD4_EM_blood_unstim.v.CD4_Nav_blood_unstim.logFC,y=CD4_EM_blood_stim.v.CD4_Nav_blood_stim.logFC)) + geom_point(aes(color=PrimeState), alpha=0.6) + geom_hline(yintercept=0, linetype="dashed") + geom_vline(xintercept=0, linetype="dashed") + theme_light() + scale_x_continuous(limits=c(-10,10)) + scale_y_continuous(limits=c(-10,10)) + theme(legend.position = 'bottom') + 
  xlab("RNA log2FC (Mem Unstim/Naive Unstim)") + ylab("RNA log2FC (Mem Stim/Naive Stim)") + labs(color="Primed Peak(s) Near Gene") + guides(alpha=F) + scale_color_manual(values=c("TRUE"="khaki3", "FALSE"="grey41"))

ggMarginal(p, groupColour = T, groupFill = T)
ggsave(here(output_dir, "CD4_EM_primed_peak_expression_nologFCcut.pdf"), height=4.5, width=5)
```

Interesting! When you filter for logFC>1 and FDR significant there is a definite skewing towards the positive direction in both unstim and stim states! Need to back it up with statisitcs though.

How does this compare with other peak types?

```{r}
pMem <- ggplot(RNA_sub2, aes(x=CD4_EM_blood_unstim.v.CD4_Nav_blood_unstim.logFC,y=CD4_EM_blood_stim.v.CD4_Nav_blood_stim.logFC)) + geom_point(aes(color=MemState),alpha=0.6) + geom_hline(yintercept=0, linetype="dashed") + geom_vline(xintercept=0, linetype="dashed") + theme_light() + scale_x_continuous(limits=c(-10,10)) + scale_y_continuous(limits=c(-10,10)) + theme(legend.position = 'bottom') + 
  xlab("RNA log2FC (Mem Unstim/Naive Unstim)") + ylab("RNA log2FC (Mem Stim/Naive Stim)") + labs(color="Memory Peak(s) Near Gene") + guides(alpha=F) + scale_color_manual(values=c("TRUE"="mediumpurple", "FALSE"="grey41"))
ggMarginal(pMem, groupColour = T, groupFill = T)
ggsave(here(output_dir, "CD4_EM_mem_peak_expression_nologFCcut.pdf"), height=4.5, width=5)
```

```{r}
pInd <- ggplot(RNA_sub2, aes(x=CD4_EM_blood_unstim.v.CD4_Nav_blood_unstim.logFC,y=CD4_EM_blood_stim.v.CD4_Nav_blood_stim.logFC)) + geom_point(aes(color=IndState), alpha=0.6) + geom_hline(yintercept=0, linetype="dashed") + geom_vline(xintercept=0, linetype="dashed") + theme_light() + scale_x_continuous(limits=c(-10,10)) + scale_y_continuous(limits=c(-10,10)) + theme(legend.position = 'bottom') + 
  xlab("RNA log2FC (Mem Unstim/Naive Unstim)") + ylab("RNA log2FC (Mem Stim/Naive Stim)") + labs(color="Induced Peak(s) Near Gene") + guides(alpha=F) + scale_color_manual(values=c("TRUE"="#C05746", "FALSE"="grey41"))
ggMarginal(pInd, groupColour = T, groupFill = T)
ggsave(here(output_dir, "CD4_EM_ind_peak_expression_nologFCcut.pdf"), height=4.5, width=5)
```

```{r}
pConsv <- ggplot(RNA_sub2, aes(x=CD4_EM_blood_unstim.v.CD4_Nav_blood_unstim.logFC,y=CD4_EM_blood_stim.v.CD4_Nav_blood_stim.logFC)) + geom_point(aes(color=ConsvState, alpha=0.6)) + geom_hline(yintercept=0, linetype="dashed") + geom_vline(xintercept=0, linetype="dashed") + theme_light() + scale_x_continuous(limits=c(-10,10)) + scale_y_continuous(limits=c(-10,10)) + theme(legend.position = 'bottom') + 
  xlab("RNA log2FC (Mem Unstim/Naive Unstim)") + ylab("RNA log2FC (Mem Stim/Naive Stim)") + labs(color="Conserved Peak(s) Near Gene") + guides(alpha=F) + scale_color_manual(values=c("TRUE"="orange4", "FALSE"="grey41"))
ggMarginal(pConsv, groupColour = T, groupFill = T)
ggsave(here(output_dir, "CD4_EM_consv_peak_expression_nologFCcut.pdf"), height=4.5, width=5)
```

```{r}
pNav <- ggplot(RNA_sub2, aes(x=CD4_EM_blood_unstim.v.CD4_Nav_blood_unstim.logFC,y=CD4_EM_blood_stim.v.CD4_Nav_blood_stim.logFC)) + geom_point(aes(color=NavState, alpha=0.6)) + geom_hline(yintercept=0, linetype="dashed") + geom_vline(xintercept=0, linetype="dashed") + theme_light() + scale_x_continuous(limits=c(-10,10)) + scale_y_continuous(limits=c(-10,10)) + theme(legend.position = 'bottom') + 
  xlab("RNA log2FC (Mem Unstim/Naive Unstim)") + ylab("RNA log2FC (Mem Stim/Naive Stim)") + labs(color="Naive Peak(s) Near Gene") + guides(alpha=F) + scale_color_manual(values=c("TRUE"="#2A9D8F", "FALSE"="grey41"))
ggMarginal(pNav, groupColour = T, groupFill = T)
ggsave(here(output_dir, "CD4_EM_nav_peak_expression_nologFCcut.pdf"), height=4.5, width=5)
```

```{r}
pOther <- ggplot(RNA_sub2, aes(x=CD4_EM_blood_unstim.v.CD4_Nav_blood_unstim.logFC,y=CD4_EM_blood_stim.v.CD4_Nav_blood_stim.logFC)) + geom_point(aes(color=OtherState, alpha=0.6)) + geom_hline(yintercept=0, linetype="dashed") + geom_vline(xintercept=0, linetype="dashed") + theme_light() + scale_x_continuous(limits=c(-10,10)) + scale_y_continuous(limits=c(-10,10)) + theme(legend.position = 'bottom') + 
  xlab("RNA log2FC (Mem Unstim/Naive Unstim)") + ylab("RNA log2FC (Mem Stim/Naive Stim)") + labs(color="Uncategorized Peak(s) Near Gene") + guides(alpha=F) + scale_color_manual(values=c("TRUE"="navy", "FALSE"="grey41"))
ggMarginal(pOther, groupColour = T, groupFill = T)
ggsave(here(output_dir, "CD4_EM_other_peak_expression_nologFCcut.pdf"), height=4.5, width=5)
```

Show same as scatter above but with simpler plots
```{r}
ggplot(RNA_sub2, aes(x=PrimeState,y=CD4_EM_blood_unstim.v.CD4_Nav_blood_unstim.logFC)) + geom_jitter(alpha=0.4) + geom_violin(aes(fill=PrimeState)) + geom_jitter(alpha=0.4)

ggplot(RNA_sub2, aes(x=PrimeState,y=CD4_EM_blood_stim.v.CD4_Nav_blood_stim.logFC)) + geom_jitter(alpha=0.4) + geom_violin(aes(fill=PrimeState)) + geom_jitter(alpha=0.4)

ggplot(RNA_sub2, aes(x=IndState,y=CD4_EM_blood_unstim.v.CD4_Nav_blood_unstim.logFC)) + geom_jitter(alpha=0.4) + geom_violin(aes(fill=IndState)) + geom_jitter(alpha=0.4)

ggplot(RNA_sub2, aes(x=IndState,y=CD4_EM_blood_stim.v.CD4_Nav_blood_stim.logFC)) + geom_jitter(alpha=0.4) + geom_violin(aes(fill=IndState)) + geom_jitter(alpha=0.4)

ggplot(RNA_sub2, aes(x=PrimeState,y=CD4_EM_blood_unstim.v.CD4_Nav_blood_unstim.logFC)) + geom_boxplot(aes(fill=PrimeState))
ggplot(RNA_sub2, aes(x=PrimeState,y=CD4_EM_blood_stim.v.CD4_Nav_blood_stim.logFC)) + geom_boxplot(aes(fill=PrimeState))
```

I'd like to combine the boxplots of both unstim and stim logFC values and group by different peaktypes. I'll have to reformat the dataframe to do so I though.

```{r}
RNA_sub3 <- select(RNA_sub2, SYMBOL, CD4_EM_blood_unstim.v.CD4_Nav_blood_unstim.logFC,CD4_EM_blood_stim.v.CD4_Nav_blood_stim.logFC, PrimeState, IndState, MemState,ConsvState,NavState,OtherState) %>% pivot_longer(cols=ends_with(".logFC"), names_to="Stimulation", values_to="logFC")
RNA_sub3$Stimulation <- factor(RNA_sub3$Stimulation, levels=c("CD4_EM_blood_unstim.v.CD4_Nav_blood_unstim.logFC",
                                                            "CD4_EM_blood_stim.v.CD4_Nav_blood_stim.logFC"), ordered = T)
RNA_sub3$Stimulation <- recode(RNA_sub3$Stimulation,
                          CD4_EM_blood_unstim.v.CD4_Nav_blood_unstim.logFC="unstim",                                CD4_EM_blood_stim.v.CD4_Nav_blood_stim.logFC="stim"
                          )

```

```{r}
ggplot(RNA_sub3, aes(x=Stimulation, y=logFC)) + geom_boxplot(aes(fill=PrimeState))
ggplot(RNA_sub3, aes(x=Stimulation, y=logFC)) + geom_boxplot(aes(fill=IndState))
```

Use ggpubr to add stats

```{r}
library(ggpubr)

comparisons <- list(c("FALSE","TRUE"))
# RNA_sub3$PrimeState <- as.numeric(RNA_sub3$PrimeState)
# RNA_sub3$PrimeState <- factor(RNA_sub3$PrimeState)

pubPrime <- ggboxplot(RNA_sub3, x="Stimulation", y="logFC", color="PrimeState", palette = c("TRUE"="khaki3", "FALSE"="grey41"))
pubPrime 
pubInd <- ggboxplot(RNA_sub3, x="Stimulation", y="logFC", color="IndState", palette = c("TRUE"="#C05746", "FALSE"="grey41"))
pubInd

pubNav <- ggboxplot(RNA_sub3, x="Stimulation", y="logFC", color="NavState", palette = c("TRUE"="#2A9D8F", "FALSE"="grey41"))
pubNav
```

```{r}
# pubPrimeUnstim <- ggboxplot(subset(RNA_sub3, Stimulation=="unstim"), x="PrimeState", y="logFC", color="PrimeState", palette = c("TRUE"="khaki3", "FALSE"="grey41"))
# pubPrimeUnstim + stat_compare_means(comparisons = comparisons, type)

ggboxplot(RNA_sub3, x="PrimeState", y="logFC", color="PrimeState", palette = c("TRUE"="khaki3", "FALSE"="grey41"), facet.by = "Stimulation",
          outlier.shape = NA, notch=T) + stat_compare_means(aes(label = paste0("p = ", ..p.format..)), vjust=0.5,method="wilcox.test") + scale_y_continuous(limits=c(-3,3))
ggsave(here(output_dir, "CD4_EM_PrimePeak_statslogFCboxplots.pdf"), height=4.5, width=3)

ggboxplot(RNA_sub3, x="IndState", y="logFC", color="IndState", palette = c("TRUE"="#C05746", "FALSE"="grey41"), facet.by = "Stimulation",
          outlier.shape = NA, notch=T) + stat_compare_means(aes(label = paste0("p = ", ..p.format..)), vjust=0.5,method="wilcox.test") + scale_y_continuous(limits=c(-3,3))
ggsave(here(output_dir, "CD4_EM_StimPeak_statslogFCboxplots.pdf"), height=4.5, width=3)

ggboxplot(RNA_sub3, x="NavState", y="logFC", color="NavState", palette = c("TRUE"="#2A9D8F", "FALSE"="grey41"), facet.by = "Stimulation",
          outlier.shape = NA, notch=T) + stat_compare_means(aes(label = paste0("p = ", ..p.format..)), vjust=0.5) + scale_y_continuous(limits=c(-3,3))
ggsave(here(output_dir, "CD4_EM_NaivePeak_statslogFCboxplots.pdf"), height=4.5, width=3)

ggboxplot(RNA_sub3, x="MemState", y="logFC", color="MemState", palette = c("TRUE"="mediumpurple", "FALSE"="grey41"), facet.by = "Stimulation",
          outlier.shape = NA, notch=T) + stat_compare_means(aes(label = paste0("p = ", ..p.format..)), vjust=0.5) + scale_y_continuous(limits=c(-3,3))
ggsave(here(output_dir, "CD4_EM_MemoryPeak_statslogFCboxplots.pdf"), height=4.5, width=3)

ggboxplot(RNA_sub3, x="ConsvState", y="logFC", color="ConsvState", palette = c("TRUE"="orange4", "FALSE"="grey41"), facet.by = "Stimulation",
          outlier.shape = NA, notch=T) + stat_compare_means(aes(label = paste0("p = ", ..p.format..)), vjust=0.5) + scale_y_continuous(limits=c(-3,3))
ggsave(here(output_dir, "CD4_EM_ConsvPeak_statslogFCboxplots.pdf"), height=4.5, width=3)

ggboxplot(RNA_sub3, x="OtherState", y="logFC", color="OtherState", palette = c("TRUE"="navy", "FALSE"="grey41"), facet.by = "Stimulation",
          outlier.shape = NA, notch=T) + stat_compare_means(aes(label = paste0("p = ", ..p.format..)), vjust=0.5) + scale_y_continuous(limits=c(-3,3))
ggsave(here(output_dir, "CD4_EM_OtherPeak_statslogFCboxplots.pdf"), height=4.5, width=3)

```

Running just the stats
```{r}
var.test(logFC ~ PrimeState, data=subset(RNA_sub3, Stimulation=="unstim"))
var.test(logFC ~ PrimeState, data=subset(RNA_sub3, Stimulation=="stim"))

var.test(logFC ~ IndState, data=subset(RNA_sub3, Stimulation=="unstim"))
var.test(logFC ~ IndState, data=subset(RNA_sub3, Stimulation=="stim"))


shapiro.test(subset(RNA_sub3, Stimulation=="unstim" & PrimeState==T)$logFC)
```

Permutation based
```{r}
library(coin)
independence_test(logFC ~ PrimeState,
                  data=subset(RNA_sub3, Stimulation=="unstim"))
independence_test(logFC ~ PrimeState,
                  data=subset(RNA_sub3, Stimulation=="stim"))
independence_test(logFC ~ IndState,
                  data=subset(RNA_sub3, Stimulation=="unstim"))
independence_test(logFC ~ IndState,
                  data=subset(RNA_sub3, Stimulation=="stim"))
```



It looks like the variance for the unstim points is not equal so I'll switch to the Welches T test

```{r}
hist(subset(RNA_sub3, Stimulation=="unstim" & PrimeState==T)$logFC)
hist(subset(RNA_sub3, Stimulation=="unstim" & PrimeState==F)$logFC)
hist(subset(RNA_sub3, Stimulation=="unstim" & IndState==T)$logFC)
hist(subset(RNA_sub3, Stimulation=="unstim" & IndState==F)$logFC)
```



```{r}
# pubPrime
# ggsave(here(output_dir, "CD4_EM_PrimePeak_logFCboxplots.pdf"), height=4.5, width=3)
# pubInd
# ggsave(here(output_dir, "CD4_EM_StimPeak_logFCboxplots.pdf"), height=4.5, width=3)
# pubNav
# ggsave(here(output_dir, "CD4_EM_NavPeak_logFCboxplots.pdf"), height=4.5, width=3)
```

Can't get the stats to work right but the plots look okay.


### Let's see if these shifts in distributions can be quantified 

Using cumulative distribution plots

Primed in unstim
```{r}
ecdf_unstim_prime <- ggplot(RNA_sub2, aes(x=CD4_EM_blood_unstim.v.CD4_Nav_blood_unstim.logFC)) + stat_ecdf(geom="point", aes(color=PrimeState)) + geom_vline(xintercept = 1, linetype="dashed")+ geom_vline(xintercept = -1, linetype="dashed") + scale_x_continuous(limits=c(-10,10)) + theme_light() + scale_color_manual(values=c("TRUE"="khaki3", "FALSE"="grey41")) + labs(x="log2FC Unstimulated (Mem/Nav)")
ecdf_unstim_prime
```

Primed in Stim

```{r}
ecdf_stim_prime <- ggplot(RNA_sub2, aes(x=CD4_EM_blood_stim.v.CD4_Nav_blood_stim.logFC)) + stat_ecdf(geom="point", aes(color=PrimeState)) + geom_vline(xintercept = 1, linetype="dashed")+ geom_vline(xintercept = -1, linetype="dashed") + scale_x_continuous(limits=c(-10,10)) + theme_light() + scale_color_manual(values=c("TRUE"="khaki3", "FALSE"="grey41")) + labs(x="log2FC Stimulated (Mem/Nav)")
ecdf_stim_prime
```

Induced in unstim

```{r}
ecdf_unstim_ind<-ggplot(RNA_sub2, aes(x=CD4_EM_blood_unstim.v.CD4_Nav_blood_unstim.logFC)) + stat_ecdf(geom="point", aes(color=IndState))+ geom_vline(xintercept = 1, linetype="dashed")+ geom_vline(xintercept = -1, linetype="dashed") + scale_x_continuous(limits=c(-10,10)) + theme_light() + scale_color_manual(values=c("TRUE"="#C05746", "FALSE"="grey41")) + labs(x="log2FC Unstimulated (Mem/Nav)")
ecdf_stim_ind<-ggplot(RNA_sub2, aes(x=CD4_EM_blood_stim.v.CD4_Nav_blood_stim.logFC)) + stat_ecdf(geom="point", aes(color=IndState)) + geom_vline(xintercept = 1, linetype="dashed")+ geom_vline(xintercept = -1, linetype="dashed") + scale_x_continuous(limits=c(-10,10)) + theme_light() + scale_color_manual(values=c("TRUE"="#C05746", "FALSE"="grey41")) + labs(x="log2FC Stimulated (Mem/Nav)")

ecdf_unstim_ind
ecdf_stim_ind
```

Mem

```{r}
ecdf_unstim_mem<-ggplot(RNA_sub2, aes(x=CD4_EM_blood_unstim.v.CD4_Nav_blood_unstim.logFC)) + stat_ecdf(geom="point", aes(color=MemState))+ geom_vline(xintercept = 1, linetype="dashed")+ geom_vline(xintercept = -1, linetype="dashed") + scale_x_continuous(limits=c(-10,10)) + theme_light() + scale_color_manual(values=c("TRUE"="mediumpurple", "FALSE"="grey41")) + labs(x="log2FC Unstimulated (Mem/Nav)")
ecdf_stim_mem<-ggplot(RNA_sub2, aes(x=CD4_EM_blood_stim.v.CD4_Nav_blood_stim.logFC)) + stat_ecdf(geom="point", aes(color=MemState)) + geom_vline(xintercept = 1, linetype="dashed")+ geom_vline(xintercept = -1, linetype="dashed") + scale_x_continuous(limits=c(-10,10)) + theme_light() + scale_color_manual(values=c("TRUE"="mediumpurple", "FALSE"="grey41")) + labs(x="log2FC Stimulated (Mem/Nav)")
ecdf_unstim_mem
ecdf_stim_mem
```

Nav

```{r}
ecdf_unstim_nav<-ggplot(RNA_sub2, aes(x=CD4_EM_blood_unstim.v.CD4_Nav_blood_unstim.logFC)) + stat_ecdf(geom="point", aes(color=NavState))+ geom_vline(xintercept = 1, linetype="dashed")+ geom_vline(xintercept = -1, linetype="dashed") + scale_x_continuous(limits=c(-10,10)) + theme_light() + scale_color_manual(values=c("TRUE"="#2A9D8F", "FALSE"="grey41")) + labs(x="log2FC Unstimulated (Mem/Nav)")
ecdf_stim_nav<-ggplot(RNA_sub2, aes(x=CD4_EM_blood_stim.v.CD4_Nav_blood_stim.logFC)) + stat_ecdf(geom="point", aes(color=NavState)) + geom_vline(xintercept = 1, linetype="dashed")+ geom_vline(xintercept = -1, linetype="dashed") + scale_x_continuous(limits=c(-10,10)) + theme_light() + scale_color_manual(values=c("TRUE"="#2A9D8F", "FALSE"="grey41")) + labs(x="log2FC Stimulated (Mem/Nav)")
ecdf_unstim_nav
ecdf_stim_nav
```

Conserved

```{r}
ecdf_unstim_Consv<-ggplot(RNA_sub2, aes(x=CD4_EM_blood_unstim.v.CD4_Nav_blood_unstim.logFC)) + stat_ecdf(geom="point", aes(color=ConsvState))+ geom_vline(xintercept = 1, linetype="dashed")+ geom_vline(xintercept = -1, linetype="dashed") + scale_x_continuous(limits=c(-10,10)) + theme_light() + scale_color_manual(values=c("TRUE"="orange4", "FALSE"="grey41")) + labs(x="log2FC Unstimulated (Mem/Nav)")
ecdf_stim_Consv<-ggplot(RNA_sub2, aes(x=CD4_EM_blood_stim.v.CD4_Nav_blood_stim.logFC)) + stat_ecdf(geom="point", aes(color=ConsvState)) + geom_vline(xintercept = 1, linetype="dashed")+ geom_vline(xintercept = -1, linetype="dashed") + scale_x_continuous(limits=c(-10,10)) + theme_light() + scale_color_manual(values=c("TRUE"="orange4", "FALSE"="grey41")) + labs(x="log2FC Stimulated (Mem/Nav)")
ecdf_unstim_Consv
ecdf_stim_Consv
```

Let's try and plot all the peaktypes on one graph

```{r}
RNA_prime <- subset(RNA_sub2, PrimeState==T)
RNA_mem <- subset(RNA_sub2, MemState==T)
RNA_ind <- subset(RNA_sub2, IndState==T)
RNA_consv <- subset(RNA_sub2, ConsvState==T)
RNA_nav <- subset(RNA_sub2, NavState==T)

multi_peak_unstim <- ggplot(RNA_prime, aes(x=CD4_EM_blood_unstim.v.CD4_Nav_blood_unstim.logFC)) + stat_ecdf(geom="point", color="khaki3") + stat_ecdf(data=RNA_mem, geom="point", color="mediumpurple") + stat_ecdf(data=RNA_ind, geom="point", color="#C05746") + stat_ecdf(data=RNA_consv, geom="point", color="orange4") + stat_ecdf(data=RNA_nav, geom="point", color="#2A9D8F") + geom_vline(xintercept = 1, linetype="dashed")+ geom_vline(xintercept = -1, linetype="dashed") + scale_x_continuous(limits=c(-5,5)) + theme_light() + scale_color_manual(values=c("TRUE"="khaki3", "FALSE"="grey41")) + labs(x="log2FC Unstimulated (Mem/Nav)")
multi_peak_unstim 

multi_peak_stim <- ggplot(RNA_prime, aes(x=CD4_EM_blood_stim.v.CD4_Nav_blood_stim.logFC)) + stat_ecdf(geom="point", color="khaki3") + stat_ecdf(data=RNA_mem, geom="point", color="mediumpurple") + stat_ecdf(data=RNA_ind, geom="point", color="#C05746") + stat_ecdf(data=RNA_consv, geom="point", color="orange4") + stat_ecdf(data=RNA_nav, geom="point", color="#2A9D8F") + geom_vline(xintercept = 1, linetype="dashed")+ geom_vline(xintercept = -1, linetype="dashed") + scale_x_continuous(limits=c(-5,5)) + theme_light() + scale_color_manual(values=c("TRUE"="khaki3", "FALSE"="grey41")) + labs(x="log2FC Stimulated (Mem/Nav)")
multi_peak_stim 
```


```{r}
library(ggpubr)

#ggarrange(ecdf_unstim_prime, ecdf_stim_prime, common.legend = T)
#ggarrange(ecdf_unstim_ind, ecdf_stim_ind, common.legend = T)

ggarrange(ecdf_unstim_prime, ecdf_stim_prime, ecdf_unstim_ind, ecdf_stim_ind, 
          ncol=2, nrow=2)
ggsave(here(output_dir, "ecdf_PrimeInd_combined.pdf"), height=6, width=10)

ggarrange(ecdf_unstim_mem, ecdf_stim_mem, ecdf_unstim_Consv, ecdf_stim_Consv, 
          ncol=2, nrow=2)
ggsave(here(output_dir, "ecdf_MemConsv_combined.pdf"), height=6, width=10)


ggarrange(multi_peak_unstim, multi_peak_stim)
ggsave(here(output_dir, "ecdf_multi.pdf"), height=5, width=10)
```



```{r}
library(UpSetR)

States <- select(RNA_sub2, ends_with("State"))
States_int <- as.data.frame(apply(States, c(2), function(x) as.integer(x)))
upset(States_int)
```



Same plot without coloring
```{r}
ggplot(RNA_sub2, aes(x=CD4_EM_blood_unstim.v.CD4_Nav_blood_unstim.logFC,y=CD4_EM_blood_stim.v.CD4_Nav_blood_stim.logFC)) + geom_point() + geom_hline(yintercept=0, linetype="dashed") + geom_vline(xintercept=0, linetype="dashed") + theme_light() + scale_x_continuous(limits=c(-10,10)) + scale_y_continuous(limits=c(-10,10)) + theme(legend.position = 'bottom') + 
  xlab("RNA log2FC (Mem Unstim/Naive Unstim)") + ylab("RNA log2FC (Mem Stim/Naive Stim)") + guides(alpha=F)
ggsave(filename=here(output_dir, "CD4_EM_nocolor_expression.pdf"), height=4, width=4.5)
```


### What about with the filtering for distance to TSS?

```{r}
p2 <- ggplot(RNA_sub2, aes(x=CD4_EM_blood_unstim.v.CD4_Nav_blood_unstim.logFC,y=CD4_EM_blood_stim.v.CD4_Nav_blood_stim.logFC)) + geom_point(aes(color=PrimeState_Strict, alpha=0.6)) + geom_hline(yintercept=0, linetype="dashed") + geom_vline(xintercept=0, linetype="dashed") + theme_light() + scale_x_continuous(limits=c(-10,10)) + scale_y_continuous(limits=c(-10,10)) + theme(legend.position = 'bottom') + 
  xlab("RNA log2FC (Mem Unstim/Naive Unstim)") + ylab("RNA log2FC (Mem Stim/Naive Stim)") + labs(color="Primed Peak(s) Near Gene") + guides(alpha=F)

ggMarginal(p2, groupColour = T, groupFill = T)
```

Pretty similar.

### How about my memory expression modules? Where do they fall?

Setting up RNA dataframe with expression module categories

```{r}
vars <- c("CD4_Nav_blood_unstim.v.CD4_CM_blood_unstim",
          "CD4_Nav_blood_stim.v.CD4_CM_blood_stim",
          "CD4_Nav_blood_unstim.v.CD4_EM_blood_unstim",
          "CD4_Nav_blood_stim.v.CD4_EM_blood_stim",
          "CD8_Nav_blood_unstim.v.CD8_CM_blood_unstim",
          "CD8_Nav_blood_stim.v.CD8_CM_blood_stim",
          "CD8_Nav_blood_unstim.v.CD8_EM_blood_unstim",
          "CD8_Nav_blood_stim.v.CD8_EM_blood_stim",
          "CD8_Nav_blood_unstim.v.CD8_EMRA_blood_unstim",
          "CD8_Nav_blood_stim.v.CD8_EMRA_blood_stim"
          )
sig_vars <- paste(vars, ".sig", sep="")
fc_vars <- paste(vars, ".logFC", sep="")


CD4_CM <- select(RNA_sub, SYMBOL, ENTREZID, !!fc_vars[1], !!fc_vars[2], !!sig_vars[1], !!sig_vars[2])
names(CD4_CM) <- c("SYMBOL", "ENTREZID","Unstim_logFC", "Stim_logFC", "Unstim_sig", "Stim_sig")
CD4_CM$sigtotal <- select(CD4_CM, Unstim_sig, Stim_sig) %>% rowSums() %>% factor()
CD4_CM <- mutate(CD4_CM, Unstim_logFC = -Unstim_logFC,
                 Stim_logFC = -Stim_logFC,
                 diff=abs(Unstim_logFC-Stim_logFC)
                 ) 

CD4_EM<- select(RNA_sub,SYMBOL, ENTREZID, !!fc_vars[3], !!fc_vars[4], !!sig_vars[3], !!sig_vars[4])
names(CD4_EM) <- c("SYMBOL", "ENTREZID","Unstim_logFC", "Stim_logFC", "Unstim_sig", "Stim_sig")
CD4_EM$sigtotal <- select(CD4_EM, Unstim_sig, Stim_sig) %>% rowSums() %>% factor()
CD4_EM<- mutate(CD4_EM, Unstim_logFC = -Unstim_logFC,
                 Stim_logFC = -Stim_logFC,
                diff=abs(Unstim_logFC-Stim_logFC)
                )

CD8_CM <- select(RNA_sub,SYMBOL, ENTREZID, !!fc_vars[5], !!fc_vars[6], !!sig_vars[5], !!sig_vars[6])
names(CD8_CM) <- c("SYMBOL", "ENTREZID","Unstim_logFC", "Stim_logFC", "Unstim_sig", "Stim_sig")
CD8_CM$sigtotal <- select(CD8_CM, Unstim_sig, Stim_sig) %>% rowSums() %>% factor()
CD8_CM <- mutate(CD8_CM, Unstim_logFC = -Unstim_logFC,
                 Stim_logFC = -Stim_logFC,
                 diff=abs(Unstim_logFC-Stim_logFC)
                )

CD8_EM <- select(RNA_sub,SYMBOL, ENTREZID, !!fc_vars[7], !!fc_vars[8], !!sig_vars[7], !!sig_vars[8] )
names(CD8_EM) <- c("SYMBOL","ENTREZID", "Unstim_logFC", "Stim_logFC", "Unstim_sig", "Stim_sig")
CD8_EM$sigtotal <- select(CD8_EM, Unstim_sig, Stim_sig) %>% rowSums() %>% factor()
CD8_EM <- mutate(CD8_EM, Unstim_logFC = -Unstim_logFC,
                 Stim_logFC = -Stim_logFC,
                 diff=abs(Unstim_logFC-Stim_logFC)
                 )

CD8_EMRA <- select(RNA_sub,SYMBOL, ENTREZID, !!fc_vars[9], !!fc_vars[10], !!sig_vars[9], !!sig_vars[10])
names(CD8_EMRA) <- c("SYMBOL","ENTREZID", "Unstim_logFC", "Stim_logFC", "Unstim_sig", "Stim_sig")
CD8_EMRA$sigtotal <- select(CD8_EMRA, Unstim_sig, Stim_sig) %>% rowSums() %>% factor()
CD8_EMRA <- mutate(CD8_EMRA, Unstim_logFC = -Unstim_logFC,
                 Stim_logFC = -Stim_logFC,
                 diff=abs(Unstim_logFC-Stim_logFC)
                 )

folds <- list(CD4_CM=CD4_CM,
              CD4_EM=CD4_EM,
              CD8_EM=CD8_EM,
              CD8_CM=CD8_CM,
              CD8_EMRA=CD8_EMRA)

##Function definition
sig_class <- function(X){
  X$sig <- vector(mode="character", length=length(X$SYMBOL))
  
  for (i in 1:length(X$SYMBOL)){
    if (X$sigtotal[i]==1 & X[i,5]==TRUE){
      X$sig[i]=names(X)[5]
    } else if (X$sigtotal[i]==1 & X[i,6]==TRUE){
      X$sig[i]=names(X)[6]
    } else if (X$sigtotal[i]==2) {
      X$sig[i]="Both"
    } else if (X$sigtotal[i]==0) {
      X$sig[i]="None"
    } else {
      X$sig[i]=NA
    }
  }
  X$sig <- as.factor(X$sig)
  return(X)
}

#Mapping through folds list
folds <- map(folds, sig_class)

sig_full <- function(x){
  x <- mutate(x, 
         sig_full = case_when(
              sig == "Unstim_sig" & Unstim_logFC > 0 ~ "Mem_expressed_up",
              sig == "Unstim_sig" & Unstim_logFC < 0 ~ "Mem_expressed_dn",
              sig == "Stim_sig" & Stim_logFC > 0 ~ "Mem_induced_up",
              sig == "Stim_sig" & Stim_logFC < 0 ~ "Mem_induced_dn",
              sig == "Both" & Unstim_logFC > 0 & Stim_logFC >0 ~ "Mem_Aug_up_up",
              sig == "Both" & Unstim_logFC < 0 & Stim_logFC >0 ~ "Mem_Aug_dn_up",
              sig == "Both" & Unstim_logFC < 0 & Stim_logFC <0 ~ "Mem_Aug_dn_dn",
              sig == "Both" & Unstim_logFC > 0 & Stim_logFC <0 ~ "Mem_Aug_up_dn"
                                ),
         PrimeState=case_when(
                              SYMBOL %in% primed$Gene.Name ~ TRUE,
                              ELSE ~ FALSE
                                        ),
         PrimeState_Strict = case_when(
                             SYMBOL %in% primed_strict$Gene.Name ~ TRUE,
                             ELSE ~ FALSE
                             )
  )
  x$sig_full <- as.factor(x$sig_full)
  return(x)
}

folds <- map(folds, sig_full)
str(folds)
```

```{r}
library(ggrepel)
r = 4

CD4_EM_primed <- subset(folds$CD4_EM, PrimeState==T)

ggplot(CD4_EM_primed, aes(x=Unstim_logFC,y=Stim_logFC)) + geom_point(aes(color=sig_full)) + geom_hline(yintercept=0, linetype="dashed") + geom_vline(xintercept=0, linetype="dashed") + theme_light() +
# geom_text_repel(aes(label=if_else((Unstim_logFC<=1 & abs(Stim_logFC)>5.5) |(Unstim_logFC>5 & abs(Stim_logFC)<=1), as.character(SYMBOL),"")), min.segment.length = 0)  +
# geom_text_repel(aes(label=if_else(
#                         sqrt(abs(Unstim_logFC)) + sqrt(abs(Stim_logFC)) > r | 
#                         (Unstim_logFC<=1 & abs(Stim_logFC)>5.5) | 
#                         (Unstim_logFC>5 & abs(Stim_logFC)<=1) |
#                         SYMBOL=="MNDA" | SYMBOL =="EOMES",
#                         as.character(SYMBOL),""
#                                 )),
#                 min.segment.length = 0,
#                 size=3
#   ) +
  geom_text_repel(aes(label=if_else(
                        sqrt(abs(Unstim_logFC)) + sqrt(abs(Stim_logFC)) > r,
                        as.character(SYMBOL),""
                                )),
                min.segment.length = 0,
                size=3,
                color="purple"
  ) +
  geom_text_repel(aes(label=if_else(
                        (Unstim_logFC<=1 & abs(Stim_logFC)>5.5),
                        as.character(SYMBOL),""
                                )),
                min.segment.length = 0,
                size=3,
                color="orange"
  ) +
  geom_text_repel(aes(label=if_else(
                        (Unstim_logFC>5 & abs(Stim_logFC)<=1),
                        as.character(SYMBOL),""
                                )),
                min.segment.length = 0,
                size=3,
                color="firebrick"
  ) +
  geom_text_repel(aes(label=if_else(
                        SYMBOL=="MNDA",
                        as.character(SYMBOL),""
                                )),
                min.segment.length = 0,
                size=3,
                color="cornflower blue"
  ) +
    geom_text_repel(aes(label=if_else(
                        SYMBOL=="EOMES",
                        as.character(SYMBOL),""
                                )),
                min.segment.length = 0,
                size=3,
                color="purple"
  ) +
scale_x_continuous(limits=c(-10,10)) + scale_y_continuous(limits=c(-10,10)) +
scale_color_manual(name="DEG Category", values=c("None"="grey", "Mem_induced_dn"="light blue", "Mem_induced_up"="orange", "Mem_Aug_dn_dn"="violet","Mem_Aug_up_up"="purple", "Mem_Aug_dn_up"="purple","Mem_Aug_up_dn"="purple","Mem_expressed_dn"="cornflower blue", "Mem_expressed_up"="firebrick")) +
theme(legend.position = 'right') + 
  xlab("RNA log2FC (Mem Unstim/Naive Unstim)") + ylab("RNA log2FC (Mem Stim/Naive Stim)") + guides(alpha=F)
ggsave(here(output_dir, "CD4_EM_cats_expression2.pdf"), height=4, width=6.5)
```

```{r}
library(tidyverse)
all <- summary(folds$CD4_EM$sig_full)
prime <- summary(CD4_EM_primed$sig_full)

test <- prime/all
test

CD4_per_prime <- data.frame(all=all, prime=prime, percent=test, cat=names(test)) %>% subset(cat!="NA's" & cat!="Mem_Aug_up_dn")
ggplot(CD4_per_prime, aes(x=cat, y=percent)) + geom_col(aes(fill=cat)) +
  scale_fill_manual(name="DEG Category", values=c("None"="grey", "Mem_induced_dn"="light blue", "Mem_induced_up"="orange", "Mem_Aug_dn_dn"="violet","Mem_Aug_up_up"="purple", "Mem_Aug_dn_up"="purple","Mem_Aug_up_dn"="purple","Mem_expressed_dn"="cornflower blue", "Mem_expressed_up"="firebrick")) + theme_classic() + theme(axis.text.x=element_blank()) + ylab("Percent Genes with Primed ATAC Peak") + xlab("DEG Expression Category") 
ggsave(filename=here(output_dir, "CD4_Expression_Primed_barplot.pdf"), width=4.5, height=4)
```

So it looks like nearly 80% of the upregulated augmented category genes have a primed peak in cis. It's closer to 50% for the other two categories.

Can I look at this for all the subsets now?

```{r}
library(data.table)

primed_lst <- list()
per_prime_lst <- list()

for (i in 1:length(folds)){
  primed_lst[[i]] <- subset(folds[[i]], PrimeState==T)
  
  all_tmp <- summary(folds[[i]]$sig_full)
  prime_tmp <- summary(primed_lst[[i]]$sig_full)
  
  test_tmp <- prime_tmp/all_tmp
  
  per_prime_lst[[i]] <- data.frame(all=all_tmp, prime=prime_tmp, percent=test_tmp, cat=names(test_tmp), subset=rep(names(folds)[[i]], length(test_tmp))) 
  
}

per_prime_df <- rbindlist(per_prime_lst)
per_prime_df <- subset(per_prime_df, cat!="NA's" & cat!="Mem_Aug_up_dn" & cat!="Mem_Aug_dn_up")

per_prime_df$subset <- factor(per_prime_df$subset, levels=c("CD4_CM", "CD4_EM", "CD8_CM", "CD8_EM", "CD8_EMRA"), ordered = F)
```

```{r}
ggplot(per_prime_df, aes(x=cat, y=percent)) + geom_col(aes(fill=cat)) +
  facet_wrap(vars(subset)) +
  scale_fill_manual(name="DEG Category", values=c("None"="grey", "Mem_induced_dn"="light blue", "Mem_induced_up"="orange", "Mem_Aug_dn_dn"="violet","Mem_Aug_up_up"="purple", "Mem_Aug_dn_up"="purple","Mem_Aug_up_dn"="purple","Mem_expressed_dn"="cornflower blue", "Mem_expressed_up"="firebrick")) + theme_light() + theme(axis.text.x=element_blank()) + ylab("Percent Genes with Primed ATAC Peak") + xlab("DEG Expression Category")
ggsave(filename=here(output_dir, "all_Expression_Primed_barplot.pdf"), width=7, height=4)
```



Does this effect change by memory subset?

```{r}
RNA_sub_CD4_CM <- subset(RNA_sub, abs(CD4_Nav_blood_unstim.v.CD4_CM_blood_unstim.logFC)>FCcutoff & CD4_Nav_blood_unstim.v.CD4_CM_blood_unstim.fdr <FDRcutoff |
                            abs(CD4_Nav_blood_stim.v.CD4_CM_blood_stim.logFC)>FCcutoff & CD4_Nav_blood_stim.v.CD4_CM_blood_stim.fdr <FDRcutoff) %>%
            mutate(CD4_CM_blood_unstim.v.CD4_Nav_blood_unstim.logFC=-CD4_Nav_blood_unstim.v.CD4_CM_blood_unstim.logFC,
                   CD4_CM_blood_stim.v.CD4_Nav_blood_stim.logFC=-CD4_Nav_blood_stim.v.CD4_CM_blood_stim.logFC
                   )

p_CD4_CM <- ggplot(RNA_sub_CD4_CM, aes(x=CD4_CM_blood_unstim.v.CD4_Nav_blood_unstim.logFC,y=CD4_CM_blood_stim.v.CD4_Nav_blood_stim.logFC)) + geom_point(aes(color=PrimeState, alpha=0.6)) + geom_hline(yintercept=0, linetype="dashed") + geom_vline(xintercept=0, linetype="dashed") + theme_light() + scale_x_continuous(limits=c(-10,10)) + scale_y_continuous(limits=c(-10,10)) + theme(legend.position = 'bottom') + 
  xlab("RNA log2FC (Mem Unstim/Naive Unstim)") + ylab("RNA log2FC (Mem Stim/Naive Stim)") + labs(color="Primed Peak(s) Near Gene") + guides(alpha=F)

ggMarginal(p_CD4_CM, groupColour = T, groupFill = T)
ggsave(here(output_dir, "CD4_CM_primed_peak_expression.pdf"), height=4.5, width=5)
```

```{r}
RNA_sub_CD8_CM <- subset(RNA_sub, abs(CD8_Nav_blood_unstim.v.CD8_CM_blood_unstim.logFC)>FCcutoff & CD8_Nav_blood_unstim.v.CD8_CM_blood_unstim.fdr <FDRcutoff |
                            abs(CD8_Nav_blood_stim.v.CD8_CM_blood_stim.logFC)>FCcutoff & CD8_Nav_blood_stim.v.CD8_CM_blood_stim.fdr <FDRcutoff) %>%
            mutate(CD8_CM_blood_unstim.v.CD8_Nav_blood_unstim.logFC=-CD8_Nav_blood_unstim.v.CD8_CM_blood_unstim.logFC,
                   CD8_CM_blood_stim.v.CD8_Nav_blood_stim.logFC=-CD8_Nav_blood_stim.v.CD8_CM_blood_stim.logFC
                   )

p_CD8_CM <- ggplot(RNA_sub_CD8_CM, aes(x=CD8_CM_blood_unstim.v.CD8_Nav_blood_unstim.logFC,y=CD8_CM_blood_stim.v.CD8_Nav_blood_stim.logFC)) + geom_point(aes(color=PrimeState, alpha=0.6)) + geom_hline(yintercept=0, linetype="dashed") + geom_vline(xintercept=0, linetype="dashed") + theme_light() + scale_x_continuous(limits=c(-10,10)) + scale_y_continuous(limits=c(-10,10)) + theme(legend.position = 'bottom') + 
  xlab("RNA log2FC (Mem Unstim/Naive Unstim)") + ylab("RNA log2FC (Mem Stim/Naive Stim)") + labs(color="Primed Peak(s) Near Gene") + guides(alpha=F)

ggMarginal(p_CD8_CM, groupColour = T, groupFill = T)
ggsave(here(output_dir, "CD8_CM_primed_peak_expression.pdf"), height=4.5, width=5)
```

```{r}
RNA_sub_CD8_EM <- subset(RNA_sub, abs(CD8_Nav_blood_unstim.v.CD8_EM_blood_unstim.logFC)>FCcutoff & CD8_Nav_blood_unstim.v.CD8_EM_blood_unstim.fdr <FDRcutoff |
                            abs(CD8_Nav_blood_stim.v.CD8_EM_blood_stim.logFC)>FCcutoff & CD8_Nav_blood_stim.v.CD8_EM_blood_stim.fdr <FDRcutoff) %>%
            mutate(CD8_EM_blood_unstim.v.CD8_Nav_blood_unstim.logFC=-CD8_Nav_blood_unstim.v.CD8_EM_blood_unstim.logFC,
                   CD8_EM_blood_stim.v.CD8_Nav_blood_stim.logFC=-CD8_Nav_blood_stim.v.CD8_EM_blood_stim.logFC
                   )

p_CD8_EM <- ggplot(RNA_sub_CD8_EM, aes(x=CD8_EM_blood_unstim.v.CD8_Nav_blood_unstim.logFC,y=CD8_EM_blood_stim.v.CD8_Nav_blood_stim.logFC)) + geom_point(aes(color=PrimeState, alpha=0.6)) + geom_hline(yintercept=0, linetype="dashed") + geom_vline(xintercept=0, linetype="dashed") + theme_light() + scale_x_continuous(limits=c(-10,10)) + scale_y_continuous(limits=c(-10,10)) + theme(legend.position = 'bottom') + 
  xlab("RNA log2FC (Mem Unstim/Naive Unstim)") + ylab("RNA log2FC (Mem Stim/Naive Stim)") + labs(color="Primed Peak(s) Near Gene") + guides(alpha=F)

ggMarginal(p_CD8_EM, groupColour = T, groupFill = T)
ggsave(here(output_dir, "CD8_EM_primed_peak_expression.pdf"), height=4.5, width=5)
```

```{r}
RNA_sub_CD8_EMRA <- subset(RNA_sub, abs(CD8_Nav_blood_unstim.v.CD8_EMRA_blood_unstim.logFC)>FCcutoff & CD8_Nav_blood_unstim.v.CD8_EMRA_blood_unstim.fdr <FDRcutoff |
                            abs(CD8_Nav_blood_stim.v.CD8_EMRA_blood_stim.logFC)>FCcutoff & CD8_Nav_blood_stim.v.CD8_EMRA_blood_stim.fdr <FDRcutoff) %>%
            mutate(CD8_EMRA_blood_unstim.v.CD8_Nav_blood_unstim.logFC=-CD8_Nav_blood_unstim.v.CD8_EMRA_blood_unstim.logFC,
                   CD8_EMRA_blood_stim.v.CD8_Nav_blood_stim.logFC=-CD8_Nav_blood_stim.v.CD8_EMRA_blood_stim.logFC
                   )

p_CD8_EMRA <- ggplot(RNA_sub_CD8_EMRA, aes(x=CD8_EMRA_blood_unstim.v.CD8_Nav_blood_unstim.logFC,y=CD8_EMRA_blood_stim.v.CD8_Nav_blood_stim.logFC)) + geom_point(aes(color=PrimeState, alpha=0.6)) + geom_hline(yintercept=0, linetype="dashed") + geom_vline(xintercept=0, linetype="dashed") + theme_light() + scale_x_continuous(limits=c(-10,10)) + scale_y_continuous(limits=c(-10,10)) + theme(legend.position = 'bottom') + 
  xlab("RNA log2FC (Mem Unstim/Naive Unstim)") + ylab("RNA log2FC (Mem Stim/Naive Stim)") + labs(color="Primed Peak(s) Near Gene") + guides(alpha=F)

ggMarginal(p_CD8_EMRA, groupColour = T, groupFill = T)
ggsave(here(output_dir, "CD8_EMRA_primed_peak_expression.pdf"), height=4.5, width=5)
```

Label some interesting points

```{r}
library(ggrepel)
p
plabel <- p + geom_text_repel(aes(label=if_else(CD4_EM_blood_stim.v.CD4_Nav_blood_stim.logFC>6 & abs(CD4_EM_blood_unstim.v.CD4_Nav_blood_unstim.logFC)<=1 & PrimeState==T, as.character(SYMBOL),"")), min.segment.length = 0) + geom_text_repel(aes(label=if_else(CD4_EM_blood_stim.v.CD4_Nav_blood_stim.logFC<=1 & abs(CD4_EM_blood_unstim.v.CD4_Nav_blood_unstim.logFC)>5 & PrimeState==T, as.character(SYMBOL),"")), min.segment.length = 0)

ggMarginal(plabel, groupColour = T, groupFill = T)
ggsave(here(output_dir, "CD4_EM_primed_peak_expression_label.pdf"), height=4.5, width=5)
```


Let's do some boxplots

```{r}
my_comp <- list(c('FALSE', 'TRUE'))

p_CD4_CM_box <- ggboxplot(RNA_sub_CD4_CM, x="PrimeState", y="CD4_CM_blood_unstim.v.CD4_Nav_blood_unstim.logFC", color="PrimeState")
p_CD4_CM_vln <- ggviolin(RNA_sub_CD4_CM, x="PrimeState", y="CD4_CM_blood_unstim.v.CD4_Nav_blood_unstim.logFC", color="PrimeState")
p_CD4_CM_box + stat_compare_means(comparisons = my_comp, label = "p.signif") + labs(x="Primed Peak(s) Near Gene", y="RNA log2FC (Mem Unstim/Naive Unstim)")
p_CD4_CM_box2 <- ggboxplot(RNA_sub_CD4_CM, x="PrimeState", y="CD4_CM_blood_stim.v.CD4_Nav_blood_stim.logFC", color="PrimeState")
p_CD4_CM_box2 + stat_compare_means(comparisons = my_comp, label = "p.signif") + labs(x="Primed Peak(s) Near Gene", y="RNA log2FC (Mem Stim/Naive Stim)")

ggarrange(p_CD4_CM_box + stat_compare_means(comparisons = my_comp, label = "p.signif") + labs(x="Primed Peak(s) Near Gene", y="RNA log2FC (Mem Unstim/Naive Unstim)"),
          p_CD4_CM_box2 + stat_compare_means(comparisons = my_comp, label = "p.signif") + labs(x="Primed Peak(s) Near Gene", y="RNA log2FC (Mem Stim/Naive Stim)"), common.legend = T)
ggsave(here(output_dir, "CD4_CM_boxplots_logfC.pdf"), height=4, width=6)

p_CD4_CM_vln + stat_compare_means(comparisons = my_comp, label = "p.signif") + labs(x="Primed Peak(s) Near Gene", y="RNA log2FC (Mem Stim/Naive Stim)")
```

```{r}
p_CD4_EM_box <- ggboxplot(RNA_sub2, x="PrimeState", y="CD4_EM_blood_unstim.v.CD4_Nav_blood_unstim.logFC", color="PrimeState")
p_CD4_EM_box + stat_compare_means(comparisons = my_comp, label = "p.signif") + labs(x="Primed Peak(s) Near Gene", y="RNA log2FC (Mem Unstim/Naive Unstim)")
p_CD4_EM_box2 <- ggboxplot(RNA_sub2, x="PrimeState", y="CD4_EM_blood_stim.v.CD4_Nav_blood_stim.logFC", color="PrimeState")
p_CD4_EM_box2 + stat_compare_means(comparisons = my_comp, label = "p.signif") + labs(x="Primed Peak(s) Near Gene", y="RNA log2FC (Mem Stim/Naive Stim)")

ggarrange(p_CD4_EM_box + stat_compare_means(comparisons = my_comp, label = "p.signif") + labs(x="Primed Peak(s) Near Gene", y="RNA log2FC (Mem Unstim/Naive Unstim)"),
          p_CD4_EM_box2 + stat_compare_means(comparisons = my_comp, label = "p.signif") + labs(x="Primed Peak(s) Near Gene", y="RNA log2FC (Mem Stim/Naive Stim)"), common.legend = T)
ggsave(here(output_dir, "CD4_EM_boxplots_logfC.pdf"), height=4, width=6)
```

Combining multiple subsets into one boxplot

```{r}
RNA_sub3 <- select(RNA_sub, ENTREZID, SYMBOL, ends_with(".rpkm"), starts_with("CD4_Nav_blood_unstim.v.CD4_EM_blood_unstim"),
                   starts_with("CD4_Nav_blood_unstim.v.CD4_CM_blood_unstim"),
                   starts_with("CD8_Nav_blood_unstim.v.CD8_CM_blood_unstim"),
                   starts_with("CD8_Nav_blood_unstim.v.CD8_EM_blood_unstim"),
                   starts_with("CD8_Nav_blood_unstim.v.CD8_EMRA_blood_unstim"),
                   starts_with("CD4_Nav_blood_stim.v.CD4_CM_blood_stim"),
                   starts_with("CD4_Nav_blood_stim.v.CD4_EM_blood_stim"),
                   starts_with("CD8_Nav_blood_stim.v.CD8_CM_blood_stim"),
                   starts_with("CD8_Nav_blood_stim.v.CD8_EM_blood_stim"),
                   starts_with("CD8_Nav_blood_stim.v.CD8_EMRA_blood_stim"),
                   sigAny,
                   PrimeState,
                   PrimeState_Strict
                   )
RNA_sub3_logFC <-  transmute(RNA_sub3, CD4_CM_blood_unstim.logFC=-CD4_Nav_blood_unstim.v.CD4_CM_blood_unstim.logFC,
         CD4_EM_blood_unstim.logFC=-CD4_Nav_blood_unstim.v.CD4_EM_blood_unstim.logFC,
         CD8_CM_blood_unstim.logFC=-CD8_Nav_blood_unstim.v.CD8_CM_blood_unstim.logFC,
         CD8_EM_blood_unstim.logFC=-CD8_Nav_blood_unstim.v.CD8_EM_blood_unstim.logFC,
         CD8_EMRA_blood_unstim.logFC=-CD8_Nav_blood_unstim.v.CD8_EMRA_blood_unstim.logFC,
         CD4_CM_blood_stim.logFC=-CD4_Nav_blood_stim.v.CD4_CM_blood_stim.logFC,
         CD4_EM_blood_stim.logFC=-CD4_Nav_blood_stim.v.CD4_EM_blood_stim.logFC,
         CD8_CM_blood_stim.logFC=-CD8_Nav_blood_stim.v.CD8_CM_blood_stim.logFC,
         CD8_EM_blood_stim.logFC=-CD8_Nav_blood_stim.v.CD8_EM_blood_stim.logFC,
         CD8_EMRA_blood_stim.logFC=-CD8_Nav_blood_stim.v.CD8_EMRA_blood_stim.logFC
  )
RNA_sub3 <- select(RNA_sub3, -contains(".logFC"))
RNA_sub3 <- bind_cols(RNA_sub3, RNA_sub3_logFC)
keep2 <-RNA_sub3[apply(RNA_sub3[,grepl(".sig", names(RNA_sub3))],1,function(x) any(x==TRUE)),'SYMBOL']
RNA_sub3 = RNA_sub3[RNA_sub3$SYMBOL%in%keep2$SYMBOL,]

rm(RNA_sub3_logFC)
gc()

RNA_sub3 <- RNA_sub3 %>% pivot_longer(cols=ends_with(".logFC"), names_to="Subset", values_to="logFC") %>% separate(Subset, into=c("Celltype", "Mem_Sub", "Tissue", "Treatment"), sep="_") %>% mutate(cellSubtype = paste(Celltype, Mem_Sub, sep="_"))
RNA_sub3$Treatment <- gsub(".logFC", "", RNA_sub3$Treatment)

RNA_sub3$Treatment <- factor(RNA_sub3$Treatment, levels=c("unstim", "stim"), ordered=T)

ggboxplot(RNA_sub3, x="cellSubtype", y="logFC", color="PrimeState", facet.by="Treatment") + stat_compare_means(aes(group = PrimeState), label = "p.format") + labs(color="Primed Peak(s) Near Gene")
ggsave(filename=here(output_dir, "PrimState_multiSubset_boxplot.pdf"), height = 4, width = 12)
```

It seems like the difference in expression for genes with prime peaks is present regardless of subtyype. 

Let's check that those p-values are accurate:

```{r}
RNA_sub2_primed <- subset(RNA_sub2, PrimeState==T)
RNA_sub2_Notprimed <- subset(RNA_sub2, PrimeState==F)

wilcox.test(RNA_sub2_primed$CD4_EM_blood_unstim.v.CD4_Nav_blood_unstim.logFC, RNA_sub2_Notprimed$CD4_EM_blood_unstim.v.CD4_Nav_blood_unstim.logFC)
```

Seems to be legit

Let's heatmap these primed genes

```{r}
library(ComplexHeatmap)

rpkm<- cbind(RNA_sub[,c("SYMBOL")], RNA_sub[,grep("(CD8|CD4)_((CM)|(EM)|(EMRA)|(N))(Stim|).rpkm", colnames(RNA_sub))])
colnames(rpkm )[1] <- "SYMBOL"
mean_rpkm  <- pivot_longer(rpkm , 
                          cols=ends_with('.rpkm'),
                          names_to=c("Sample", "Celltype", "Subtype"),
                          names_pattern= "RNA_(.*)_(.*)_(.*).rpkm",
                          values_to="rpkm"
                          ) %>%
  mutate(cellSubtype=paste(Celltype, Subtype, sep="_")) %>%
  group_by(SYMBOL, cellSubtype) %>%
  summarise(mean_rpkm=mean(rpkm)) %>%
  pivot_wider(names_from=cellSubtype, values_from=mean_rpkm) %>%
  select(SYMBOL, CD4_N, CD4_NStim, CD4_CM, CD4_CMStim, CD4_EM, CD4_EMStim, CD8_N, CD8_NStim, CD8_CM, CD8_CMStim, CD8_EM, CD8_EMStim, CD8_EMRA, CD8_EMRAStim)

mean_rpkm_mtx <- as.matrix(mean_rpkm[,-1])
rownames(mean_rpkm_mtx) <- mean_rpkm$SYMBOL

mean_rpkm_mtx_z <- t(apply(mean_rpkm_mtx, 1, scale))
colnames(mean_rpkm_mtx_z) <- colnames(mean_rpkm_mtx)

mean_rpkm_mtx_prime <- mean_rpkm_mtx_z[rownames(mean_rpkm_mtx_z) %in% RNA_sub$SYMBOL[RNA_sub$PrimeState==T] & rownames(mean_rpkm_mtx_z) %in% unique(RNA_sub3$SYMBOL),]
dim(mean_rpkm_mtx_prime)

Heatmap(mean_rpkm_mtx_prime, 
        show_row_names = F)
```

I'm not sure what this is telling me that is of any use =(

```{r}
subset(RNA_sub2, SYMBOL=="PDCD1") %>% select(SYMBOL, ends_with("State"))
subset(CD4_EM_primed, SYMBOL=="PDCD1")
```

